
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://chasemc.github.io/developing/python/classes/">
      
      
        <link rel="prev" href="../setup/">
      
      
        <link rel="next" href="../testing/">
      
      
      <link rel="icon" href="../../../media/single.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.44">
    
    
      
        <title>Classes - SocialGene</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.0253249f.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-YSRY3GE6B5"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-YSRY3GE6B5",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-YSRY3GE6B5",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#socialgene.base.socialgene" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="SocialGene" class="md-header__button md-logo" aria-label="SocialGene" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SocialGene
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Classes
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/socialgene/socialgene.github.io" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="SocialGene" class="md-nav__button md-logo" aria-label="SocialGene" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    SocialGene
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/socialgene/socialgene.github.io" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../system_requirements/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    System Requirements
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../ultraquickstart/ultraquickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ultra-Quickstart!
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Building a Database (Nextflow)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Building a Database (Nextflow)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/inputs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inputs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/outputs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Outputs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/running_the_pipeline/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running the Workflow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/examples/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Examples
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_7" >
        
          
          <label class="md-nav__link" for="__nav_4_7" id="__nav_4_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Advanced
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_7">
            <span class="md-nav__icon md-icon"></span>
            Advanced
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/docker_images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Docker Images
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Nodes and Relationships
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Nodes and Relationships
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nodes_and_relationships/about_the_database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Identifiers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nodes_and_relationships/schema/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schema
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Interacting with a Database (Neo4j)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Interacting with a Database (Neo4j)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../neo4j/database_launch/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Launch
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../neo4j/database_backups/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Backups
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../neo4j/interacting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interacting
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../neo4j/slurm/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running the Database with Slurm
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" checked>
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developing
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Developing
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7_1" checked>
        
          
          <label class="md-nav__link" for="__nav_7_1" id="__nav_7_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_7_1">
            <span class="md-nav__icon md-icon"></span>
            Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Classes
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Classes
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.socialgene" class="md-nav__link">
    <span class="md-ellipsis">
      socialgene
    </span>
  </a>
  
    <nav class="md-nav" aria-label="socialgene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene" class="md-nav__link">
    <span class="md-ellipsis">
      SocialGene
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SocialGene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.__add__" class="md-nav__link">
    <span class="md-ellipsis">
      __add__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.annotate" class="md-nav__link">
    <span class="md-ellipsis">
      annotate
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.annotate_proteins_with_hmmscan" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_proteins_with_hmmscan
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.annotate_proteins_with_neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_proteins_with_neo4j
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.drop_all_non_protein_features" class="md-nav__link">
    <span class="md-ellipsis">
      drop_all_non_protein_features
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.eat_pickle" class="md-nav__link">
    <span class="md-ellipsis">
      eat_pickle
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.export_all_domains_as_tsv" class="md-nav__link">
    <span class="md-ellipsis">
      export_all_domains_as_tsv
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.ferment_pickle" class="md-nav__link">
    <span class="md-ellipsis">
      ferment_pickle
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.fill_given_locus_range" class="md-nav__link">
    <span class="md-ellipsis">
      fill_given_locus_range
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.filter_proteins" class="md-nav__link">
    <span class="md-ellipsis">
      filter_proteins
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.hydrate_from_proteins" class="md-nav__link">
    <span class="md-ellipsis">
      hydrate_from_proteins
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.hydrate_protein_info" class="md-nav__link">
    <span class="md-ellipsis">
      hydrate_protein_info
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_assembly" class="md-nav__link">
    <span class="md-ellipsis">
      table_assembly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_assembly_to_locus" class="md-nav__link">
    <span class="md-ellipsis">
      table_assembly_to_locus
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_assembly_to_taxid" class="md-nav__link">
    <span class="md-ellipsis">
      table_assembly_to_taxid
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_loci" class="md-nav__link">
    <span class="md-ellipsis">
      table_loci
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_locus_to_protein" class="md-nav__link">
    <span class="md-ellipsis">
      table_locus_to_protein
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_protein_ids" class="md-nav__link">
    <span class="md-ellipsis">
      table_protein_ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_protein_to_go" class="md-nav__link">
    <span class="md-ellipsis">
      table_protein_to_go
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.write_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.write_n_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_n_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.write_table" class="md-nav__link">
    <span class="md-ellipsis">
      write_table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio" class="md-nav__link">
    <span class="md-ellipsis">
      molbio
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Assembly">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Assembly.feature_uid_set" class="md-nav__link">
    <span class="md-ellipsis">
      feature_uid_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Assembly.add_locus" class="md-nav__link">
    <span class="md-ellipsis">
      add_locus
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain" class="md-nav__link">
    <span class="md-ellipsis">
      Domain
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Domain">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain.domain_within_threshold" class="md-nav__link">
    <span class="md-ellipsis">
      domain_within_threshold
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain.__hash__" class="md-nav__link">
    <span class="md-ellipsis">
      __hash__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature" class="md-nav__link">
    <span class="md-ellipsis">
      Feature
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Feature">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      __eq__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.__hash__" class="md-nav__link">
    <span class="md-ellipsis">
      __hash__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.feature_is_protein" class="md-nav__link">
    <span class="md-ellipsis">
      feature_is_protein
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.FeatureCollection" class="md-nav__link">
    <span class="md-ellipsis">
      FeatureCollection
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FeatureCollection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.FeatureCollection.features_sorted_by_midpoint" class="md-nav__link">
    <span class="md-ellipsis">
      features_sorted_by_midpoint
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.FeatureCollection.add_feature" class="md-nav__link">
    <span class="md-ellipsis">
      add_feature
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Location" class="md-nav__link">
    <span class="md-ellipsis">
      Location
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Location">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Location.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Locus" class="md-nav__link">
    <span class="md-ellipsis">
      Locus
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio" class="md-nav__link">
    <span class="md-ellipsis">
      Molbio
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Molbio">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio.add_assembly" class="md-nav__link">
    <span class="md-ellipsis">
      add_assembly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio.add_protein" class="md-nav__link">
    <span class="md-ellipsis">
      add_protein
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio.get_all_feature_uids" class="md-nav__link">
    <span class="md-ellipsis">
      get_all_feature_uids
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein" class="md-nav__link">
    <span class="md-ellipsis">
      Protein
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Protein">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.domain_vector" class="md-nav__link">
    <span class="md-ellipsis">
      domain_vector
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.add_domain" class="md-nav__link">
    <span class="md-ellipsis">
      add_domain
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.filter_domains" class="md-nav__link">
    <span class="md-ellipsis">
      filter_domains
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence" class="md-nav__link">
    <span class="md-ellipsis">
      ProteinSequence
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ProteinSequence">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence.seq_len" class="md-nav__link">
    <span class="md-ellipsis">
      seq_len
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence.all_attributes" class="md-nav__link">
    <span class="md-ellipsis">
      all_attributes
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Taxonomy" class="md-nav__link">
    <span class="md-ellipsis">
      Taxonomy
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../testing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Testing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Auto-Generated Reference
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Precomputed Databases
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Precomputed Databases
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8_1" >
        
          
          <label class="md-nav__link" for="__nav_8_1" id="__nav_8_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Version: 2023_v0.4.1
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_8_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8_1">
            <span class="md-nav__icon md-icon"></span>
            Version: 2023_v0.4.1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../precomputed_databases/2023_v0.4.1/general/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    general
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8_1_2" >
        
          
          <label class="md-nav__link" for="__nav_8_1_2" id="__nav_8_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    aws
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_8_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8_1_2">
            <span class="md-nav__icon md-icon"></span>
            aws
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../precomputed_databases/2023_v0.4.1/aws/aws/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../precomputed_databases/2023_v0.4.1/aws/aws_instructions/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    instructions
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../precomputed_databases/2023_v0.4.1/aws/local_use/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    local use
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../precomputed_databases/2023_v0.4.1/dryad/dryad/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dryad
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../presentations/presentations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Presentations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hasher/hasher/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    hasher
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../feedback/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Give Feedback
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.socialgene" class="md-nav__link">
    <span class="md-ellipsis">
      socialgene
    </span>
  </a>
  
    <nav class="md-nav" aria-label="socialgene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene" class="md-nav__link">
    <span class="md-ellipsis">
      SocialGene
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SocialGene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.__add__" class="md-nav__link">
    <span class="md-ellipsis">
      __add__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.annotate" class="md-nav__link">
    <span class="md-ellipsis">
      annotate
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.annotate_proteins_with_hmmscan" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_proteins_with_hmmscan
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.annotate_proteins_with_neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_proteins_with_neo4j
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.drop_all_non_protein_features" class="md-nav__link">
    <span class="md-ellipsis">
      drop_all_non_protein_features
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.eat_pickle" class="md-nav__link">
    <span class="md-ellipsis">
      eat_pickle
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.export_all_domains_as_tsv" class="md-nav__link">
    <span class="md-ellipsis">
      export_all_domains_as_tsv
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.ferment_pickle" class="md-nav__link">
    <span class="md-ellipsis">
      ferment_pickle
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.fill_given_locus_range" class="md-nav__link">
    <span class="md-ellipsis">
      fill_given_locus_range
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.filter_proteins" class="md-nav__link">
    <span class="md-ellipsis">
      filter_proteins
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.hydrate_from_proteins" class="md-nav__link">
    <span class="md-ellipsis">
      hydrate_from_proteins
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.hydrate_protein_info" class="md-nav__link">
    <span class="md-ellipsis">
      hydrate_protein_info
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_assembly" class="md-nav__link">
    <span class="md-ellipsis">
      table_assembly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_assembly_to_locus" class="md-nav__link">
    <span class="md-ellipsis">
      table_assembly_to_locus
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_assembly_to_taxid" class="md-nav__link">
    <span class="md-ellipsis">
      table_assembly_to_taxid
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_loci" class="md-nav__link">
    <span class="md-ellipsis">
      table_loci
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_locus_to_protein" class="md-nav__link">
    <span class="md-ellipsis">
      table_locus_to_protein
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_protein_ids" class="md-nav__link">
    <span class="md-ellipsis">
      table_protein_ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.table_protein_to_go" class="md-nav__link">
    <span class="md-ellipsis">
      table_protein_to_go
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.write_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.write_n_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      write_n_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.socialgene.SocialGene.write_table" class="md-nav__link">
    <span class="md-ellipsis">
      write_table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio" class="md-nav__link">
    <span class="md-ellipsis">
      molbio
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Assembly">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Assembly.feature_uid_set" class="md-nav__link">
    <span class="md-ellipsis">
      feature_uid_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Assembly.add_locus" class="md-nav__link">
    <span class="md-ellipsis">
      add_locus
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain" class="md-nav__link">
    <span class="md-ellipsis">
      Domain
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Domain">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain.domain_within_threshold" class="md-nav__link">
    <span class="md-ellipsis">
      domain_within_threshold
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain.__hash__" class="md-nav__link">
    <span class="md-ellipsis">
      __hash__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Domain.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature" class="md-nav__link">
    <span class="md-ellipsis">
      Feature
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Feature">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      __eq__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.__hash__" class="md-nav__link">
    <span class="md-ellipsis">
      __hash__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Feature.feature_is_protein" class="md-nav__link">
    <span class="md-ellipsis">
      feature_is_protein
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.FeatureCollection" class="md-nav__link">
    <span class="md-ellipsis">
      FeatureCollection
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FeatureCollection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.FeatureCollection.features_sorted_by_midpoint" class="md-nav__link">
    <span class="md-ellipsis">
      features_sorted_by_midpoint
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.FeatureCollection.add_feature" class="md-nav__link">
    <span class="md-ellipsis">
      add_feature
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Location" class="md-nav__link">
    <span class="md-ellipsis">
      Location
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Location">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Location.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Locus" class="md-nav__link">
    <span class="md-ellipsis">
      Locus
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio" class="md-nav__link">
    <span class="md-ellipsis">
      Molbio
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Molbio">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio.add_assembly" class="md-nav__link">
    <span class="md-ellipsis">
      add_assembly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio.add_protein" class="md-nav__link">
    <span class="md-ellipsis">
      add_protein
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Molbio.get_all_feature_uids" class="md-nav__link">
    <span class="md-ellipsis">
      get_all_feature_uids
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein" class="md-nav__link">
    <span class="md-ellipsis">
      Protein
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Protein">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.domain_vector" class="md-nav__link">
    <span class="md-ellipsis">
      domain_vector
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.add_domain" class="md-nav__link">
    <span class="md-ellipsis">
      add_domain
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Protein.filter_domains" class="md-nav__link">
    <span class="md-ellipsis">
      filter_domains
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence" class="md-nav__link">
    <span class="md-ellipsis">
      ProteinSequence
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ProteinSequence">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence.seq_len" class="md-nav__link">
    <span class="md-ellipsis">
      seq_len
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#socialgene.base.molbio.ProteinSequence.all_attributes" class="md-nav__link">
    <span class="md-ellipsis">
      all_attributes
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#socialgene.base.molbio.Taxonomy" class="md-nav__link">
    <span class="md-ellipsis">
      Taxonomy
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/socialgene/socialgene.github.io/edit/main/docs/developing/python/classes.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


  <h1>Classes</h1>

<div class="doc doc-object doc-module">



<h3 id="socialgene.base.socialgene" class="doc doc-heading">
            <code>socialgene.base.socialgene</code>


</h3>

    <div class="doc doc-contents first">








  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h4 id="socialgene.base.socialgene.SocialGene" class="doc doc-heading">
            <code>SocialGene</code>


</h4>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="socialgene.base.molbio.Molbio" href="#socialgene.base.molbio.Molbio">Molbio</a></code>, <code><span title="socialgene.base.compare_protein.CompareProtein">CompareProtein</span></code>, <code><span title="socialgene.parsers.sequence_parser.SequenceParser">SequenceParser</span></code>, <code><span title="socialgene.neo4j.neo4j.Neo4jQuery">Neo4jQuery</span></code>, <code><span title="socialgene.parsers.hmmer_parser.HmmerParser">HmmerParser</span></code></p>


        <p>Main class for building, sotring, working with protein and genomic info</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">SocialGene</span><span class="p">(</span><span class="n">Molbio</span><span class="p">,</span> <span class="n">CompareProtein</span><span class="p">,</span> <span class="n">SequenceParser</span><span class="p">,</span> <span class="n">Neo4jQuery</span><span class="p">,</span> <span class="n">HmmerParser</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main class for building, sotring, working with protein and genomic info&quot;&quot;&quot;</span>

    <span class="c1"># _export_table_names exports tables for the nextflow pipeline</span>
    <span class="n">_export_table_names</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;table_protein_to_go&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table_assembly_to_locus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table_loci&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table_locus_to_protein&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table_assembly&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table_assembly_to_taxid&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table_protein_ids&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_comparison</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1">########################################</span>
    <span class="c1"># Get info</span>
    <span class="c1">########################################</span>

    <span class="k">def</span> <span class="nf">get_loci_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="k">def</span> <span class="nf">get_all_gene_clusters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">j</span><span class="o">.</span><span class="n">gene_clusters</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">k</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">protein_iter</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="n">i</span>

    <span class="c1">########################################</span>
    <span class="c1"># Filter</span>
    <span class="c1">########################################</span>

    <span class="k">def</span> <span class="nf">get_protein_domains_from_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_id_list</span><span class="p">):</span>
        <span class="c1"># Search neo4j for a protein with the same hash, if it exists,</span>
        <span class="c1"># retrieve the domain info for it</span>
        <span class="n">get_protein_domains_result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">query_neo4j</span><span class="p">(</span>
            <span class="n">cypher_name</span><span class="o">=</span><span class="s2">&quot;get_protein_domains&quot;</span><span class="p">,</span>
            <span class="n">param</span><span class="o">=</span><span class="n">protein_id_list</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seq_pro_score&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;evalue&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;domain_bias&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;domain_score&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;seq_pro_bias&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;hmm_from&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;hmm_to&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;ali_from&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;ali_to&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;env_from&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="s2">&quot;env_to&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="n">get_protein_domains_result</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">get_protein_domains_result</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">domain</span> <span class="ow">in</span> <span class="n">protein</span><span class="p">[</span><span class="s2">&quot;domains&quot;</span><span class="p">]:</span>
                    <span class="n">new_dict</span> <span class="o">=</span> <span class="n">temp</span> <span class="o">|</span> <span class="n">domain</span><span class="p">[</span><span class="s2">&quot;domain_properties&quot;</span><span class="p">]</span>
                    <span class="n">new_dict</span><span class="p">[</span><span class="s2">&quot;hmm_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">domain</span><span class="p">[</span><span class="s2">&quot;hmm_id&quot;</span><span class="p">]</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;p1.uid&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">add_domain</span><span class="p">(</span>
                        <span class="o">**</span><span class="n">new_dict</span><span class="p">,</span>
                    <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;None of the input protein ids were found or had domains in the database&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">annotate_proteins_with_neo4j</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">protein_uids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="n">annotate_all</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">progress</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function `annotate_proteins_with_neo4j` takes a list of protein hash IDs, queries a database</span>
<span class="sd">        for matching proteins, and retrieves their HMM annotations.</span>

<span class="sd">        Args:</span>
<span class="sd">          protein_uids (List[str]): A list of protein hash IDs. These are unique identifiers for proteins in the database that you want to annotate.</span>
<span class="sd">          chunk_size (int): The `chunk_size` parameter determines the number of proteins to query at a time. Proteins are divided into chunks to improve efficiency when querying the database.  Defaults to 1000</span>
<span class="sd">          annotate_all (bool): The `annotate_all` parameter is a boolean flag that determines whether to annotate all proteins in the database or not. If set to `True`, all proteins in the database will be annotated. If set to `False`, you need to provide a list of protein hash IDs in the `protein_uids. Defaults to False</span>
<span class="sd">          progress (bool): The `progress` parameter is a boolean flag that determines whether or not to display a progress bar during the execution of the function. If `progress` is set to `True`, a progress bar will be displayed to track the progress of the function. If `progress` is set to `False`. Defaults to False</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">protein_uids</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">annotate_all</span><span class="p">:</span>
            <span class="n">protein_uids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_all_feature_uids</span><span class="p">()</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">protein_uids</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_uids</span><span class="p">]</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Searching database for HMM annotations of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span><span class="si">}</span><span class="s2"> proteins.&quot;</span>
        <span class="p">)</span>
        <span class="n">search_result</span> <span class="o">=</span> <span class="n">search_protein_hash</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">search_result</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No identical proteins found in the database.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># log.info(</span>
            <span class="c1">#     f&quot;{len([i for i in search_result.values() if i])} of {len(protein_uids)} searched proteins were found in the database, pulling their HMM annotations into python...&quot;</span>
            <span class="c1"># )</span>
            <span class="c1"># get the number of chunks needed to to query &quot;chunk_size&quot; proteins at a time</span>
            <span class="n">prot_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span>
            <span class="n">n_chunks</span> <span class="o">=</span> <span class="n">prot_len</span> <span class="o">//</span> <span class="n">chunk_size</span>
            <span class="k">if</span> <span class="n">n_chunks</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">n_chunks</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">n_chunks</span> <span class="o">&gt;</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">chunked_list</span> <span class="o">=</span> <span class="n">chunk_a_list_with_numpy</span><span class="p">(</span>
                    <span class="n">input_list</span><span class="o">=</span><span class="n">protein_uids</span><span class="p">,</span> <span class="n">n_chunks</span><span class="o">=</span><span class="n">n_chunks</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">chunked_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_uids</span><span class="p">]</span>
            <span class="k">del</span> <span class="n">protein_uids</span>
            <span class="k">if</span> <span class="n">progress</span><span class="p">:</span>
                <span class="k">with</span> <span class="n">Progress</span><span class="p">(</span><span class="n">transient</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">as</span> <span class="n">pg</span><span class="p">:</span>
                    <span class="n">task</span> <span class="o">=</span> <span class="n">pg</span><span class="o">.</span><span class="n">add_task</span><span class="p">(</span><span class="s2">&quot;Progress...&quot;</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="n">n_chunks</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">protein_id_list</span> <span class="ow">in</span> <span class="n">chunked_list</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">get_protein_domains_from_db</span><span class="p">(</span>
                            <span class="n">protein_id_list</span><span class="o">=</span><span class="n">protein_id_list</span>
                        <span class="p">)</span>
                        <span class="n">pg</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">task</span><span class="p">,</span> <span class="n">advance</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">protein_id_list</span> <span class="ow">in</span> <span class="n">chunked_list</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">get_protein_domains_from_db</span><span class="p">(</span><span class="n">protein_id_list</span><span class="o">=</span><span class="n">protein_id_list</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">search_result</span>

    <span class="k">def</span> <span class="nf">annotate</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">use_neo4j_precalc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">neo4j_chunk_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The `annotate` function is a convenience function that retrieves HMMER annotation for all</span>
<span class="sd">        proteins in a Socialgene object, either from a Neo4j database or by using HMMER directly.</span>

<span class="sd">        Args:</span>
<span class="sd">          use_neo4j_precalc (bool): A boolean flag indicating whether to use precalculated domain</span>
<span class="sd">        information from Neo4j for annotation. If set to True, the domains info will be retrieved from</span>
<span class="sd">        Neo4j, and proteins not found in Neo4j will be analyzed with HMMER. If set to False, all</span>
<span class="sd">        proteins will be analyzed with HMMER. Defaults to False</span>
<span class="sd">          neo4j_chunk_size (int): The `neo4j_chunk_size` parameter determines the number of proteins</span>
<span class="sd">        that will be sent to Neo4j at a time for annotation. Defaults to 1000</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">use_neo4j_precalc</span><span class="p">:</span>
            <span class="n">db_res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotate_proteins_with_neo4j</span><span class="p">(</span>
                <span class="n">annotate_all</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">chunk_size</span><span class="o">=</span><span class="n">neo4j_chunk_size</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">n_not_in_db</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">db_res</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">v</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">n_not_in_db</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">n_not_in_db</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotate_proteins_with_hmmscan</span><span class="p">(</span>
                <span class="n">protein_id_list</span><span class="o">=</span><span class="n">n_not_in_db</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">annotate_proteins_with_hmmscan</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">protein_id_list</span><span class="p">,</span> <span class="n">hmm_filepath</span><span class="p">,</span> <span class="n">cpus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Run hmmscan on Protein objects</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_id_list (list): list of protein hash ids to run hmmscan on</span>
<span class="sd">            hmm_filepath (str): path to file of HMMs</span>
<span class="sd">            cpus (int): number of parallel processes to spawn</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotating </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_id_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> proteins with HMMER&#39;s hmmscan&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">cpus</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">cpus</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cpus</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="c1"># create a list of proteins as FASTA</span>
        <span class="n">temp1</span> <span class="o">=</span> <span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">fasta_string_defline_uid</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">protein_id_list</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">temp1</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;None of the input sequences contain an amino acid sequence&quot;</span><span class="p">)</span>
            <span class="k">return</span>
        <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">()</span> <span class="k">as</span> <span class="n">temp_path</span><span class="p">:</span>
            <span class="n">hmmer_temp</span> <span class="o">=</span> <span class="n">HMMER</span><span class="p">(</span><span class="n">hmm_filepath</span><span class="o">=</span><span class="n">hmm_filepath</span><span class="p">)</span>
            <span class="n">hmmer_temp</span><span class="o">.</span><span class="n">hmmscan</span><span class="p">(</span>
                <span class="n">fasta_path</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span>
                <span class="nb">input</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp1</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(),</span>
                <span class="n">domtblout_path</span><span class="o">=</span><span class="n">temp_path</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cpus</span><span class="o">=</span><span class="n">cpus</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="c1"># parse the resulting domtblout file, saving results to the class proteins/domains</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parse_hmmout</span><span class="p">(</span><span class="n">temp_path</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">hmmsearch_or_hmmscan</span><span class="o">=</span><span class="s2">&quot;hmmscan&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_proteins_from_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">ncount</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    MATCH (p1:protein)</span>
<span class="sd">                    RETURN p1.uid as uid</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
                <span class="p">):</span>
                    <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_protein</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">i</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">])</span>
                    <span class="n">ncount</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Retrieved </span><span class="si">{</span><span class="n">ncount</span><span class="si">}</span><span class="s2"> proteins from Neo4j&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Error trying to retrieve proteins from Neo4j&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_sequences_from_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    MATCH (p1:protein)</span>
<span class="sd">                    WHERE p1.uid in $uid</span>
<span class="sd">                    RETURN p1.uid as uid, p1.sequence as sequence</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
                    <span class="n">uid</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_iter</span><span class="p">],</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">i</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error trying to retrieve domains for </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">hydrate_from_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Given a SocialGene object with proteins, retrieve from a running Neo4j database all locus and asssembly info for those proteins&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">Neo4jQuery</span><span class="o">.</span><span class="n">query_neo4j</span><span class="p">(</span>
            <span class="n">cypher_name</span><span class="o">=</span><span class="s2">&quot;retrieve_protein_locations&quot;</span><span class="p">,</span>
            <span class="n">param</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
        <span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">],</span> <span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">locus</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;loci&quot;</span><span class="p">]:</span>
                <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">add_locus</span><span class="p">(</span>
                    <span class="n">external_id</span><span class="o">=</span><span class="n">locus</span><span class="p">[</span><span class="s2">&quot;locus&quot;</span><span class="p">]</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">locus</span><span class="p">[</span><span class="s2">&quot;features&quot;</span><span class="p">]:</span>
                    <span class="n">_</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]]</span>
                        <span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">locus</span><span class="p">[</span><span class="s2">&quot;locus&quot;</span><span class="p">]]</span>
                        <span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                            <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span>
                            <span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;external_id&quot;</span><span class="p">],</span>
                            <span class="n">start</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;locus_start&quot;</span><span class="p">],</span>
                            <span class="n">end</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;locus_end&quot;</span><span class="p">],</span>
                            <span class="n">strand</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">],</span>
                        <span class="p">)</span>
                    <span class="p">)</span>

    <span class="k">def</span> <span class="nf">fill_given_locus_range</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">locus_uid</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Given a locus uid that&#39;s in the database, pull asssembly, locus, protein info for those proteins&quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
            <span class="n">assembly_uid</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                MATCH (a1:assembly)&lt;-[:ASSEMBLES_TO]-(n1:nucleotide {uid: $locus_uid})</span>
<span class="sd">                RETURN a1.uid as a_uid</span>
<span class="sd">            &quot;&quot;&quot;</span><span class="p">,</span>
                <span class="n">locus_uid</span><span class="o">=</span><span class="n">locus_uid</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">single</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">assembly_uid</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No assembly found in database&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">assembly_uid</span> <span class="o">=</span> <span class="n">assembly_uid</span><span class="o">.</span><span class="n">value</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">assembly_uid</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                MATCH (n1:nucleotide {uid: $locus_uid})</span>
<span class="sd">                RETURN n1 as nucleotide_node</span>
<span class="sd">            &quot;&quot;&quot;</span><span class="p">,</span>
                <span class="n">locus_uid</span><span class="o">=</span><span class="n">locus_uid</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">single</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">res</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">locus_uid</span><span class="si">}</span><span class="s2"> not found in database&quot;</span><span class="p">)</span>
        <span class="n">external_id</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">value</span><span class="p">()</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;external_id&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_uid</span><span class="p">]</span><span class="o">.</span><span class="n">add_locus</span><span class="p">(</span><span class="n">external_id</span><span class="o">=</span><span class="n">external_id</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_uid</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">external_id</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="nb">dict</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">value</span><span class="p">())</span>
        <span class="p">)</span>

        <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                MATCH (n1:nucleotide {uid: $locus_uid})-[e1:ENCODES]-&gt;(p1:protein)</span>
<span class="sd">                WHERE e1.start &gt;= $start AND e1.start &lt;= $end</span>
<span class="sd">                RETURN e1, p1</span>
<span class="sd">            &quot;&quot;&quot;</span><span class="p">,</span>
                <span class="n">locus_uid</span><span class="o">=</span><span class="n">locus_uid</span><span class="p">,</span>
                <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_protein</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">value</span><span class="p">()</span><span class="o">.</span><span class="n">end_node</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_uid</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">external_id</span><span class="p">]</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                    <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span>
                    <span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">value</span><span class="p">()</span><span class="o">.</span><span class="n">end_node</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">],</span>
                    <span class="o">**</span><span class="n">feature</span><span class="o">.</span><span class="n">value</span><span class="p">(),</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;assembly&quot;</span><span class="p">:</span> <span class="n">assembly_uid</span><span class="p">,</span> <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">external_id</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">_drop_all_cross_origin</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Drop likely cross-origin proteins (proteins that are longer than 100k bp)</span>
        <span class="c1"># have to remove from both gene cluster and locus</span>
        <span class="k">for</span> <span class="n">ak</span><span class="p">,</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">nk</span><span class="p">,</span> <span class="n">nv</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">ak</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">nk</span><span class="p">]</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nv</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">100000</span>
                <span class="p">}</span>
                <span class="k">for</span> <span class="n">gene_cluster</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">ak</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">nk</span><span class="p">]</span><span class="o">.</span><span class="n">gene_clusters</span><span class="p">:</span>
                    <span class="n">gene_cluster</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">{</span>
                        <span class="n">i</span>
                        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">gene_cluster</span><span class="o">.</span><span class="n">features</span>
                        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">100000</span>
                    <span class="p">}</span>

    <span class="k">def</span> <span class="nf">hydrate_protein_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Pull name (original identifier) and description of proteins from Neo4j&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">Neo4jQuery</span><span class="o">.</span><span class="n">query_neo4j</span><span class="p">(</span>
            <span class="n">cypher_name</span><span class="o">=</span><span class="s2">&quot;get_protein_info&quot;</span><span class="p">,</span>
            <span class="n">param</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
        <span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;external_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;external_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span>

    <span class="c1">########################################</span>
    <span class="c1"># File Outputs</span>
    <span class="c1">########################################</span>

    <span class="k">def</span> <span class="nf">export_all_domains_as_tsv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function exports all domains as a TSV file, sorted by protein ID and mean envelope position.</span>

<span class="sd">        Args:</span>
<span class="sd">          outpath: The `outpath` parameter is the path to the output file where the TSV (Tab-Separated</span>
<span class="sd">        Values) data will be written. It specifies the location and name of the file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_domain_counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">tsv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># sort to standardize the write order</span>
            <span class="n">ordered_prot_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="n">ordered_prot_ids</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">prot_id</span> <span class="ow">in</span> <span class="n">ordered_prot_ids</span><span class="p">:</span>
                <span class="c1"># sort to standardize the write order</span>
                <span class="k">for</span> <span class="n">domain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span>
                    <span class="n">prot_id</span>
                <span class="p">]</span><span class="o">.</span><span class="n">domain_list_sorted_by_mean_envelope_position</span><span class="p">:</span>
                    <span class="n">_domain_counter</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">_temp</span> <span class="o">=</span> <span class="p">[</span><span class="n">prot_id</span><span class="p">]</span>
                    <span class="n">_temp</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">domain</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
                    <span class="n">tsv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">_temp</span><span class="p">)</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">_domain_counter</span><span class="p">)</span><span class="si">}</span><span class="s2"> domains to </span><span class="si">{</span><span class="n">outpath</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">ferment_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function `ferment_pickle` saves a SocialGene object to a Python pickle file.</span>

<span class="sd">        Args:</span>
<span class="sd">          outpath: The `outpath` parameter is a string that represents the path where the pickled object will be saved. It should include the file name and extension. For example, &quot;/path/to/save/object.pickle&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">handle</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">HIGHEST_PROTOCOL</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">eat_pickle</span><span class="p">(</span><span class="n">inpath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The `eat_pickle` function reads a saved SocialGene pickle file from the given path and returns a SocialGene object.</span>

<span class="sd">        Args:</span>
<span class="sd">          inpath: The `inpath` parameter is a string that represents the path to the file from which we</span>
<span class="sd">        want to load the pickled object.</span>

<span class="sd">        Returns:</span>
<span class="sd">          SocialGene object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">inpath</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">filter_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hash_list</span><span class="p">:</span> <span class="n">List</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter proteins by list of hash ids</span>

<span class="sd">        Args:</span>
<span class="sd">            hash_list (List): List of protein hash identifiers</span>

<span class="sd">        Returns:</span>
<span class="sd">            Generator: generator returning tuple of length two -&gt; Generator[(str, &#39;Protein&#39;])</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">((</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">hash_list</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_external_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="k">def</span> <span class="nf">write_fasta</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">outpath</span><span class="p">,</span>
        <span class="n">external_id</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Write all proteins to a FASTA file</span>

<span class="sd">        Args:</span>
<span class="sd">            outpath (str): path of file that FASTA entries will be appended to</span>
<span class="sd">            external_id (bool, optional): Write protein identifiers as the hash (True) or the original identifier (False). Defaults to False.</span>
<span class="sd">            **kwargs: see print(open_write.__doc__)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">filepath</span><span class="o">=</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">external_id</span><span class="p">:</span>
                <span class="n">fasta_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fasta_string_defline_external_id</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">fasta_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fasta_string_defline_uid</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fasta_gen</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="s2">&quot;compression&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;compression&quot;</span><span class="p">]:</span>
                    <span class="n">handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">encode</span><span class="p">())</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">handle</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">counter</span><span class="p">)</span><span class="si">}</span><span class="s2"> proteins to </span><span class="si">{</span><span class="n">outpath</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">write_n_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outdir</span><span class="p">,</span> <span class="n">n_splits</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function `write_n_fasta` exports protein sequences split into multiple fasta files.</span>

<span class="sd">        Args:</span>
<span class="sd">          outdir: The `outdir` parameter is a string that specifies the directory where the fasta files</span>
<span class="sd">        will be saved.</span>
<span class="sd">          n_splits: The `n_splits` parameter in the `write_n_fasta` function determines the number of</span>
<span class="sd">        fasta files the protein sequences will be split into. By default, it is set to 1, meaning all</span>
<span class="sd">        the protein sequences will be written into a single fasta file. If you specify a value greater</span>
<span class="sd">        than. Defaults to 1</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># this can be done with itertools.batched in python 3.12</span>
        <span class="k">def</span> <span class="nf">split</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
            <span class="c1"># https://stackoverflow.com/a/2135920</span>
            <span class="n">k</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">),</span> <span class="n">n</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">(</span>
                <span class="n">a</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">k</span> <span class="o">+</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">m</span><span class="p">)</span> <span class="p">:</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">k</span> <span class="o">+</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">m</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="n">protein_list</span> <span class="o">=</span> <span class="n">split</span><span class="p">(</span>
            <span class="p">[</span><span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
            <span class="n">n_splits</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span>
                <span class="n">Path</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;fasta_split_</span><span class="si">{</span><span class="n">counter</span><span class="si">}</span><span class="s2">.faa&quot;</span><span class="p">),</span> <span class="o">**</span><span class="n">kwargs</span>
            <span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">protein_group</span><span class="p">:</span>
                    <span class="n">handle</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">def</span> <span class="nf">_merge_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sg_object</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function merges the proteins from another SOcialGene object into the current object.</span>

<span class="sd">        Args:</span>
<span class="sd">          sg_object: The `sg_object` parameter is an object of the same class as the current object. It</span>
<span class="sd">        represents another object that contains a collection of proteins.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">sg_object</span><span class="o">.</span><span class="n">proteins</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_merge_assemblies</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sg_object</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function `_merge_assemblies` merges assemblies and loci from `sg_object` into `self` if they</span>
<span class="sd">        do not already exist, and adds protein features to loci if they already exist.</span>

<span class="sd">        Args:</span>
<span class="sd">          sg_object: The `sg_object` parameter is an object of the same class as the current object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">assembly_k</span><span class="p">,</span> <span class="n">assembly_v</span> <span class="ow">in</span> <span class="n">sg_object</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">assembly_k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_k</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_v</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">locus_k</span><span class="p">,</span> <span class="n">locus_v</span> <span class="ow">in</span> <span class="n">assembly_v</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">locus_k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_k</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_k</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">locus_k</span><span class="p">]</span> <span class="o">=</span> <span class="n">locus_v</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="p">(</span>
                            <span class="n">sg_object</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_k</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">locus_k</span><span class="p">]</span><span class="o">.</span><span class="n">features</span>
                        <span class="p">):</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_k</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">locus_k</span><span class="p">]</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                                <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span>
                                <span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                                <span class="n">start</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                                <span class="n">end</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                                <span class="n">strand</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                            <span class="p">)</span>

    <span class="c1"># def deep_merge_with_sg1(sg_object_1, sg_object_2):</span>
    <span class="k">def</span> <span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sg_object</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function merges proteins and assemblies from two objects and appends protein comparison data</span>
<span class="sd">        to a list or dataframe.</span>

<span class="sd">        Args:</span>
<span class="sd">          sg_object: The `sg_object` parameter is an object of the same class as the current object. It</span>
<span class="sd">        represents another instance of the class that you want to add to the current instance.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_merge_proteins</span><span class="p">(</span><span class="n">sg_object</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_merge_assemblies</span><span class="p">(</span><span class="n">sg_object</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_comparison</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sg_object</span><span class="o">.</span><span class="n">protein_comparison</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">write_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">assembly</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">locus</span> <span class="ow">in</span> <span class="n">assembly</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
                    <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">),</span>
                    <span class="nb">id</span><span class="o">=</span><span class="n">locus</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
                    <span class="n">name</span><span class="o">=</span><span class="n">locus</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
                    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;A GenBank file generated by SocialGene.&quot;</span><span class="p">,</span>
                    <span class="n">dbxrefs</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Assembly:</span><span class="si">{</span><span class="n">locus</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
                <span class="p">)</span>
                <span class="c1"># Add annotation</span>
                <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">locus</span><span class="o">.</span><span class="n">features_sorted_by_midpoint</span><span class="p">:</span>
                    <span class="n">biofeat</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span>
                        <span class="n">FeatureLocation</span><span class="p">(</span>
                            <span class="n">start</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                            <span class="n">end</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                            <span class="n">strand</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="nb">type</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
                        <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span>
                            <span class="n">k</span><span class="p">:</span> <span class="n">v</span>
                            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                            <span class="k">if</span> <span class="n">v</span> <span class="ow">and</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;parent&quot;</span><span class="p">,</span> <span class="s2">&quot;protein&quot;</span><span class="p">]</span>
                        <span class="p">}</span>
                        <span class="o">|</span> <span class="p">{</span><span class="s2">&quot;translation&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span><span class="p">},</span>
                    <span class="p">)</span>
                    <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">biofeat</span><span class="p">)</span>
                <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;DNA&quot;</span>
                <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="p">)</span> <span class="k">as</span> <span class="n">h</span><span class="p">:</span>
                    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="n">record</span><span class="p">,</span>
                        <span class="n">h</span><span class="p">,</span>
                        <span class="s2">&quot;genbank&quot;</span><span class="p">,</span>
                    <span class="p">)</span>

    <span class="k">def</span> <span class="nf">drop_all_non_protein_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Drop features from all assembly/loci that aren&#39;t proteins/pseudo-proteins</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of features that were removed (if return_removed=True)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">a_v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">l_v</span> <span class="ow">in</span> <span class="n">a_v</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">l_v</span><span class="o">.</span><span class="n">drop_non_protein_features</span><span class="p">()</span>

    <span class="c1">########################################</span>
    <span class="c1"># OUTPUTS FOR NEXTFLOW</span>
    <span class="c1">########################################</span>

    <span class="k">def</span> <span class="nf">write_table</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">outdir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">tablename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">include_sequences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function writes a table to a specified output directory in TSV format, for import into Neo4j.</span>

<span class="sd">        Args:</span>
<span class="sd">          outdir (str): The `outdir` parameter is a string that specifies the directory where the output</span>
<span class="sd">        file will be saved.</span>
<span class="sd">          tablename (str): The `tablename` parameter is a string that specifies the name of the table to</span>
<span class="sd">        be written.</span>
<span class="sd">          filename (str): The `filename` parameter is an optional argument that specifies the name of</span>
<span class="sd">        the file to be written. If no `filename` is provided, the `tablename` will be used as the</span>
<span class="sd">        filename.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">filename</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">tablename</span>
        <span class="n">outpath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="n">tsv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span>
                <span class="n">handle</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_MINIMAL</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tablename</span><span class="p">)(</span><span class="n">include_sequences</span><span class="o">=</span><span class="n">include_sequences</span><span class="p">):</span>
                <span class="n">tsv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_protein_to_go</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function `table_protein_to_go` iterates through the assemblies, loci, and features of a</span>
<span class="sd">        given object, and yields tuples containing the protein hash and GO term for each feature that</span>
<span class="sd">        has GO terms, for import into Neo4j.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="c1"># not all features will have goterms so check here</span>
                    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">goterms</span><span class="p">:</span>
                        <span class="k">for</span> <span class="n">goterm</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">goterms</span><span class="p">:</span>
                            <span class="k">yield</span> <span class="p">(</span>
                                <span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                                <span class="n">goterm</span><span class="o">.</span><span class="n">removeprefix</span><span class="p">(</span><span class="s2">&quot;GO:&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_locus_to_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function `table_locus_to_protein` generates a table of protein information from each locus, for import into Neo4j.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">ak</span><span class="p">,</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">loci</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">temp_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">loci</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
                <span class="c1"># sort features by id then start to maintain consistent output</span>
                <span class="n">temp_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;uid&quot;</span><span class="p">))</span>
                <span class="n">temp_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;start&quot;</span><span class="p">))</span>
                <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">temp_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">feature_is_protein</span><span class="p">():</span>
                        <span class="k">yield</span> <span class="p">(</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">locus_tag</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">description</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">partial_on_complete_genome</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">missing_start</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">missing_stop</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">internal_stop</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">partial_in_the_middle_of_a_contig</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">missing_N_terminus</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">missing_C_terminus</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">frameshifted</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">too_short_partial_abutting_assembly_gap</span><span class="p">,</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">incomplete</span><span class="p">,</span>
                        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_protein_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_sequences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Protein hash id table for import into Neo4j&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">include_sequences</span><span class="p">:</span>
                <span class="k">yield</span> <span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="n">protein</span><span class="o">.</span><span class="n">crc64</span><span class="p">,</span> <span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">yield</span> <span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="n">protein</span><span class="o">.</span><span class="n">crc64</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_assembly_to_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Assembly to locus table for import into Neo4j</span>

<span class="sd">        Args:</span>
<span class="sd">            outdir (str, optional): Defaults to &quot;.&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">ak</span><span class="p">,</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="c1">#  [&quot;assembly&quot;, &quot;internal_locus_id&quot;]</span>
                <span class="k">yield</span> <span class="p">(</span><span class="n">ak</span><span class="p">,</span> <span class="n">v</span><span class="o">.</span><span class="n">uid</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_loci</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a table of loci information.</span>

<span class="sd">        Yields:</span>
<span class="sd">            tuple: (internal_locus_id, external_locus_id, [info])</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">_av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">locus</span> <span class="ow">in</span> <span class="n">_av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="k">yield</span> <span class="nb">tuple</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">locus</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span>
                    <span class="o">+</span> <span class="p">[</span><span class="n">locus</span><span class="o">.</span><span class="n">external_id</span><span class="p">]</span>
                    <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_assembly</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Assembly table for import into Neo4j</span>

<span class="sd">        Args:</span>
<span class="sd">            outdir (str, optional): Defaults to &quot;.&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">assembly</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="nb">tuple</span><span class="p">(</span>
                <span class="p">[</span><span class="n">assembly</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">assembly</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">table_assembly_to_taxid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Assembly table for import into Neo4j</span>

<span class="sd">        Args:</span>
<span class="sd">            outdir (str, optional): Defaults to &quot;.&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">taxid</span><span class="p">:</span>
                <span class="k">yield</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="o">.</span><span class="n">taxid</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.__add__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__add__</span><span class="p">(</span><span class="n">sg_object</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function merges proteins and assemblies from two objects and appends protein comparison data
to a list or dataframe.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sg_object</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>sg_object</code> parameter is an object of the same class as the current object. It</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>represents another instance of the class that you want to add to the current instance.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sg_object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function merges proteins and assemblies from two objects and appends protein comparison data</span>
<span class="sd">    to a list or dataframe.</span>

<span class="sd">    Args:</span>
<span class="sd">      sg_object: The `sg_object` parameter is an object of the same class as the current object. It</span>
<span class="sd">    represents another instance of the class that you want to add to the current instance.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_merge_proteins</span><span class="p">(</span><span class="n">sg_object</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_merge_assemblies</span><span class="p">(</span><span class="n">sg_object</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_comparison</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sg_object</span><span class="o">.</span><span class="n">protein_comparison</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.annotate" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate</span><span class="p">(</span><span class="n">use_neo4j_precalc</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">neo4j_chunk_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The <code>annotate</code> function is a convenience function that retrieves HMMER annotation for all
proteins in a Socialgene object, either from a Neo4j database or by using HMMER directly.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>use_neo4j_precalc</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A boolean flag indicating whether to use precalculated domain</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>information from Neo4j for annotation. If set to True, the domains info will be retrieved from
Neo4j, and proteins not found in Neo4j will be analyzed with HMMER. If set to False, all
proteins will be analyzed with HMMER. Defaults to False
  neo4j_chunk_size (int): The <code>neo4j_chunk_size</code> parameter determines the number of proteins
that will be sent to Neo4j at a time for annotation. Defaults to 1000</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">annotate</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">use_neo4j_precalc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">neo4j_chunk_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The `annotate` function is a convenience function that retrieves HMMER annotation for all</span>
<span class="sd">    proteins in a Socialgene object, either from a Neo4j database or by using HMMER directly.</span>

<span class="sd">    Args:</span>
<span class="sd">      use_neo4j_precalc (bool): A boolean flag indicating whether to use precalculated domain</span>
<span class="sd">    information from Neo4j for annotation. If set to True, the domains info will be retrieved from</span>
<span class="sd">    Neo4j, and proteins not found in Neo4j will be analyzed with HMMER. If set to False, all</span>
<span class="sd">    proteins will be analyzed with HMMER. Defaults to False</span>
<span class="sd">      neo4j_chunk_size (int): The `neo4j_chunk_size` parameter determines the number of proteins</span>
<span class="sd">    that will be sent to Neo4j at a time for annotation. Defaults to 1000</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">use_neo4j_precalc</span><span class="p">:</span>
        <span class="n">db_res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotate_proteins_with_neo4j</span><span class="p">(</span>
            <span class="n">annotate_all</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">chunk_size</span><span class="o">=</span><span class="n">neo4j_chunk_size</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">n_not_in_db</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">db_res</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">v</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">n_not_in_db</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">if</span> <span class="n">n_not_in_db</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotate_proteins_with_hmmscan</span><span class="p">(</span>
            <span class="n">protein_id_list</span><span class="o">=</span><span class="n">n_not_in_db</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.annotate_proteins_with_hmmscan" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_proteins_with_hmmscan</span><span class="p">(</span><span class="n">protein_id_list</span><span class="p">,</span> <span class="n">hmm_filepath</span><span class="p">,</span> <span class="n">cpus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Run hmmscan on Protein objects</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>protein_id_list</code>
            </td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of protein hash ids to run hmmscan on</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>hmm_filepath</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>path to file of HMMs</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cpus</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of parallel processes to spawn</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_proteins_with_hmmscan</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">protein_id_list</span><span class="p">,</span> <span class="n">hmm_filepath</span><span class="p">,</span> <span class="n">cpus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run hmmscan on Protein objects</span>

<span class="sd">    Args:</span>
<span class="sd">        protein_id_list (list): list of protein hash ids to run hmmscan on</span>
<span class="sd">        hmm_filepath (str): path to file of HMMs</span>
<span class="sd">        cpus (int): number of parallel processes to spawn</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotating </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_id_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> proteins with HMMER&#39;s hmmscan&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">cpus</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">cpus</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cpus</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="c1"># create a list of proteins as FASTA</span>
    <span class="n">temp1</span> <span class="o">=</span> <span class="p">[</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">fasta_string_defline_uid</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">protein_id_list</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">temp1</span><span class="p">:</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;None of the input sequences contain an amino acid sequence&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">()</span> <span class="k">as</span> <span class="n">temp_path</span><span class="p">:</span>
        <span class="n">hmmer_temp</span> <span class="o">=</span> <span class="n">HMMER</span><span class="p">(</span><span class="n">hmm_filepath</span><span class="o">=</span><span class="n">hmm_filepath</span><span class="p">)</span>
        <span class="n">hmmer_temp</span><span class="o">.</span><span class="n">hmmscan</span><span class="p">(</span>
            <span class="n">fasta_path</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span>
            <span class="nb">input</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp1</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(),</span>
            <span class="n">domtblout_path</span><span class="o">=</span><span class="n">temp_path</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">cpus</span><span class="o">=</span><span class="n">cpus</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># parse the resulting domtblout file, saving results to the class proteins/domains</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parse_hmmout</span><span class="p">(</span><span class="n">temp_path</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">hmmsearch_or_hmmscan</span><span class="o">=</span><span class="s2">&quot;hmmscan&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.annotate_proteins_with_neo4j" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_proteins_with_neo4j</span><span class="p">(</span><span class="n">protein_uids</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chunk_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">annotate_all</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function <code>annotate_proteins_with_neo4j</code> takes a list of protein hash IDs, queries a database
for matching proteins, and retrieves their HMM annotations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>protein_uids</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of protein hash IDs. These are unique identifiers for proteins in the database that you want to annotate.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>chunk_size</code> parameter determines the number of proteins to query at a time. Proteins are divided into chunks to improve efficiency when querying the database.  Defaults to 1000</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotate_all</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>annotate_all</code> parameter is a boolean flag that determines whether to annotate all proteins in the database or not. If set to <code>True</code>, all proteins in the database will be annotated. If set to <code>False</code>, you need to provide a list of protein hash IDs in the `protein_uids. Defaults to False</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>progress</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>progress</code> parameter is a boolean flag that determines whether or not to display a progress bar during the execution of the function. If <code>progress</code> is set to <code>True</code>, a progress bar will be displayed to track the progress of the function. If <code>progress</code> is set to <code>False</code>. Defaults to False</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_proteins_with_neo4j</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">protein_uids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">annotate_all</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">progress</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function `annotate_proteins_with_neo4j` takes a list of protein hash IDs, queries a database</span>
<span class="sd">    for matching proteins, and retrieves their HMM annotations.</span>

<span class="sd">    Args:</span>
<span class="sd">      protein_uids (List[str]): A list of protein hash IDs. These are unique identifiers for proteins in the database that you want to annotate.</span>
<span class="sd">      chunk_size (int): The `chunk_size` parameter determines the number of proteins to query at a time. Proteins are divided into chunks to improve efficiency when querying the database.  Defaults to 1000</span>
<span class="sd">      annotate_all (bool): The `annotate_all` parameter is a boolean flag that determines whether to annotate all proteins in the database or not. If set to `True`, all proteins in the database will be annotated. If set to `False`, you need to provide a list of protein hash IDs in the `protein_uids. Defaults to False</span>
<span class="sd">      progress (bool): The `progress` parameter is a boolean flag that determines whether or not to display a progress bar during the execution of the function. If `progress` is set to `True`, a progress bar will be displayed to track the progress of the function. If `progress` is set to `False`. Defaults to False</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">protein_uids</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">annotate_all</span><span class="p">:</span>
        <span class="n">protein_uids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_all_feature_uids</span><span class="p">()</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">protein_uids</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_uids</span><span class="p">]</span>
    <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Searching database for HMM annotations of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span><span class="si">}</span><span class="s2"> proteins.&quot;</span>
    <span class="p">)</span>
    <span class="n">search_result</span> <span class="o">=</span> <span class="n">search_protein_hash</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">search_result</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No identical proteins found in the database.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># log.info(</span>
        <span class="c1">#     f&quot;{len([i for i in search_result.values() if i])} of {len(protein_uids)} searched proteins were found in the database, pulling their HMM annotations into python...&quot;</span>
        <span class="c1"># )</span>
        <span class="c1"># get the number of chunks needed to to query &quot;chunk_size&quot; proteins at a time</span>
        <span class="n">prot_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span>
        <span class="n">n_chunks</span> <span class="o">=</span> <span class="n">prot_len</span> <span class="o">//</span> <span class="n">chunk_size</span>
        <span class="k">if</span> <span class="n">n_chunks</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">n_chunks</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">n_chunks</span> <span class="o">&gt;</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_uids</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">chunked_list</span> <span class="o">=</span> <span class="n">chunk_a_list_with_numpy</span><span class="p">(</span>
                <span class="n">input_list</span><span class="o">=</span><span class="n">protein_uids</span><span class="p">,</span> <span class="n">n_chunks</span><span class="o">=</span><span class="n">n_chunks</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">chunked_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_uids</span><span class="p">]</span>
        <span class="k">del</span> <span class="n">protein_uids</span>
        <span class="k">if</span> <span class="n">progress</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">Progress</span><span class="p">(</span><span class="n">transient</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">as</span> <span class="n">pg</span><span class="p">:</span>
                <span class="n">task</span> <span class="o">=</span> <span class="n">pg</span><span class="o">.</span><span class="n">add_task</span><span class="p">(</span><span class="s2">&quot;Progress...&quot;</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="n">n_chunks</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">protein_id_list</span> <span class="ow">in</span> <span class="n">chunked_list</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">get_protein_domains_from_db</span><span class="p">(</span>
                        <span class="n">protein_id_list</span><span class="o">=</span><span class="n">protein_id_list</span>
                    <span class="p">)</span>
                    <span class="n">pg</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">task</span><span class="p">,</span> <span class="n">advance</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">protein_id_list</span> <span class="ow">in</span> <span class="n">chunked_list</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">get_protein_domains_from_db</span><span class="p">(</span><span class="n">protein_id_list</span><span class="o">=</span><span class="n">protein_id_list</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">search_result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.drop_all_non_protein_features" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">drop_all_non_protein_features</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Drop features from all assembly/loci that aren't proteins/pseudo-proteins</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of features that were removed (if return_removed=True)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">drop_all_non_protein_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Drop features from all assembly/loci that aren&#39;t proteins/pseudo-proteins</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of features that were removed (if return_removed=True)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">a_v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">l_v</span> <span class="ow">in</span> <span class="n">a_v</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">l_v</span><span class="o">.</span><span class="n">drop_non_protein_features</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.eat_pickle" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">eat_pickle</span><span class="p">(</span><span class="n">inpath</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h5>


    <div class="doc doc-contents ">

        <p>The <code>eat_pickle</code> function reads a saved SocialGene pickle file from the given path and returns a SocialGene object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>inpath</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>inpath</code> parameter is a string that represents the path to the file from which we</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>want to load the pickled object.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SocialGene object</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">eat_pickle</span><span class="p">(</span><span class="n">inpath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The `eat_pickle` function reads a saved SocialGene pickle file from the given path and returns a SocialGene object.</span>

<span class="sd">    Args:</span>
<span class="sd">      inpath: The `inpath` parameter is a string that represents the path to the file from which we</span>
<span class="sd">    want to load the pickled object.</span>

<span class="sd">    Returns:</span>
<span class="sd">      SocialGene object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">inpath</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.export_all_domains_as_tsv" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">export_all_domains_as_tsv</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function exports all domains as a TSV file, sorted by protein ID and mean envelope position.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outpath</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>outpath</code> parameter is the path to the output file where the TSV (Tab-Separated</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Values) data will be written. It specifies the location and name of the file.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">export_all_domains_as_tsv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function exports all domains as a TSV file, sorted by protein ID and mean envelope position.</span>

<span class="sd">    Args:</span>
<span class="sd">      outpath: The `outpath` parameter is the path to the output file where the TSV (Tab-Separated</span>
<span class="sd">    Values) data will be written. It specifies the location and name of the file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_domain_counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">tsv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># sort to standardize the write order</span>
        <span class="n">ordered_prot_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">ordered_prot_ids</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">prot_id</span> <span class="ow">in</span> <span class="n">ordered_prot_ids</span><span class="p">:</span>
            <span class="c1"># sort to standardize the write order</span>
            <span class="k">for</span> <span class="n">domain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span>
                <span class="n">prot_id</span>
            <span class="p">]</span><span class="o">.</span><span class="n">domain_list_sorted_by_mean_envelope_position</span><span class="p">:</span>
                <span class="n">_domain_counter</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">_temp</span> <span class="o">=</span> <span class="p">[</span><span class="n">prot_id</span><span class="p">]</span>
                <span class="n">_temp</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">domain</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
                <span class="n">tsv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">_temp</span><span class="p">)</span>
    <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">_domain_counter</span><span class="p">)</span><span class="si">}</span><span class="s2"> domains to </span><span class="si">{</span><span class="n">outpath</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.ferment_pickle" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ferment_pickle</span><span class="p">(</span><span class="n">outpath</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function <code>ferment_pickle</code> saves a SocialGene object to a Python pickle file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outpath</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>outpath</code> parameter is a string that represents the path where the pickled object will be saved. It should include the file name and extension. For example, "/path/to/save/object.pickle".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">ferment_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function `ferment_pickle` saves a SocialGene object to a Python pickle file.</span>

<span class="sd">    Args:</span>
<span class="sd">      outpath: The `outpath` parameter is a string that represents the path where the pickled object will be saved. It should include the file name and extension. For example, &quot;/path/to/save/object.pickle&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">handle</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">HIGHEST_PROTOCOL</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.fill_given_locus_range" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fill_given_locus_range</span><span class="p">(</span><span class="n">locus_uid</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Given a locus uid that's in the database, pull asssembly, locus, protein info for those proteins</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fill_given_locus_range</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">locus_uid</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a locus uid that&#39;s in the database, pull asssembly, locus, protein info for those proteins&quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
        <span class="n">assembly_uid</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            MATCH (a1:assembly)&lt;-[:ASSEMBLES_TO]-(n1:nucleotide {uid: $locus_uid})</span>
<span class="sd">            RETURN a1.uid as a_uid</span>
<span class="sd">        &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">locus_uid</span><span class="o">=</span><span class="n">locus_uid</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">single</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">assembly_uid</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No assembly found in database&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">assembly_uid</span> <span class="o">=</span> <span class="n">assembly_uid</span><span class="o">.</span><span class="n">value</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">assembly_uid</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            MATCH (n1:nucleotide {uid: $locus_uid})</span>
<span class="sd">            RETURN n1 as nucleotide_node</span>
<span class="sd">        &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">locus_uid</span><span class="o">=</span><span class="n">locus_uid</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">single</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">res</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">locus_uid</span><span class="si">}</span><span class="s2"> not found in database&quot;</span><span class="p">)</span>
    <span class="n">external_id</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">value</span><span class="p">()</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;external_id&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_uid</span><span class="p">]</span><span class="o">.</span><span class="n">add_locus</span><span class="p">(</span><span class="n">external_id</span><span class="o">=</span><span class="n">external_id</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_uid</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">external_id</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">value</span><span class="p">())</span>
    <span class="p">)</span>

    <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            MATCH (n1:nucleotide {uid: $locus_uid})-[e1:ENCODES]-&gt;(p1:protein)</span>
<span class="sd">            WHERE e1.start &gt;= $start AND e1.start &lt;= $end</span>
<span class="sd">            RETURN e1, p1</span>
<span class="sd">        &quot;&quot;&quot;</span><span class="p">,</span>
            <span class="n">locus_uid</span><span class="o">=</span><span class="n">locus_uid</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
            <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_protein</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">value</span><span class="p">()</span><span class="o">.</span><span class="n">end_node</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">assembly_uid</span><span class="p">]</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">external_id</span><span class="p">]</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span>
                <span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">value</span><span class="p">()</span><span class="o">.</span><span class="n">end_node</span><span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">],</span>
                <span class="o">**</span><span class="n">feature</span><span class="o">.</span><span class="n">value</span><span class="p">(),</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;assembly&quot;</span><span class="p">:</span> <span class="n">assembly_uid</span><span class="p">,</span> <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">external_id</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.filter_proteins" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_proteins</span><span class="p">(</span><span class="n">hash_list</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Filter proteins by list of hash ids</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>hash_list</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of protein hash identifiers</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Generator</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>generator returning tuple of length two -&gt; Generator[(str, 'Protein'])</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hash_list</span><span class="p">:</span> <span class="n">List</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter proteins by list of hash ids</span>

<span class="sd">    Args:</span>
<span class="sd">        hash_list (List): List of protein hash identifiers</span>

<span class="sd">    Returns:</span>
<span class="sd">        Generator: generator returning tuple of length two -&gt; Generator[(str, &#39;Protein&#39;])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">((</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">hash_list</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.hydrate_from_proteins" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">hydrate_from_proteins</span><span class="p">()</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Given a SocialGene object with proteins, retrieve from a running Neo4j database all locus and asssembly info for those proteins</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">hydrate_from_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a SocialGene object with proteins, retrieve from a running Neo4j database all locus and asssembly info for those proteins&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">Neo4jQuery</span><span class="o">.</span><span class="n">query_neo4j</span><span class="p">(</span>
        <span class="n">cypher_name</span><span class="o">=</span><span class="s2">&quot;retrieve_protein_locations&quot;</span><span class="p">,</span>
        <span class="n">param</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">],</span> <span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">locus</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;loci&quot;</span><span class="p">]:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">add_locus</span><span class="p">(</span>
                <span class="n">external_id</span><span class="o">=</span><span class="n">locus</span><span class="p">[</span><span class="s2">&quot;locus&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">locus</span><span class="p">[</span><span class="s2">&quot;features&quot;</span><span class="p">]:</span>
                <span class="n">_</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]]</span>
                    <span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">locus</span><span class="p">[</span><span class="s2">&quot;locus&quot;</span><span class="p">]]</span>
                    <span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                        <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span>
                        <span class="n">uid</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;external_id&quot;</span><span class="p">],</span>
                        <span class="n">start</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;locus_start&quot;</span><span class="p">],</span>
                        <span class="n">end</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;locus_end&quot;</span><span class="p">],</span>
                        <span class="n">strand</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">],</span>
                    <span class="p">)</span>
                <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.hydrate_protein_info" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">hydrate_protein_info</span><span class="p">()</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Pull name (original identifier) and description of proteins from Neo4j</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">hydrate_protein_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Pull name (original identifier) and description of proteins from Neo4j&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">Neo4jQuery</span><span class="o">.</span><span class="n">query_neo4j</span><span class="p">(</span>
        <span class="n">cypher_name</span><span class="o">=</span><span class="s2">&quot;get_protein_info&quot;</span><span class="p">,</span>
        <span class="n">param</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;external_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;external_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_assembly" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_assembly</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Assembly table for import into Neo4j</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outdir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Defaults to ".".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_assembly</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assembly table for import into Neo4j</span>

<span class="sd">    Args:</span>
<span class="sd">        outdir (str, optional): Defaults to &quot;.&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">assembly</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">yield</span> <span class="nb">tuple</span><span class="p">(</span>
            <span class="p">[</span><span class="n">assembly</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">assembly</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_assembly_to_locus" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_assembly_to_locus</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Assembly to locus table for import into Neo4j</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outdir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Defaults to ".".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_assembly_to_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assembly to locus table for import into Neo4j</span>

<span class="sd">    Args:</span>
<span class="sd">        outdir (str, optional): Defaults to &quot;.&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">ak</span><span class="p">,</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1">#  [&quot;assembly&quot;, &quot;internal_locus_id&quot;]</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">ak</span><span class="p">,</span> <span class="n">v</span><span class="o">.</span><span class="n">uid</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_assembly_to_taxid" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_assembly_to_taxid</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Assembly table for import into Neo4j</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outdir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Defaults to ".".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_assembly_to_taxid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assembly table for import into Neo4j</span>

<span class="sd">    Args:</span>
<span class="sd">        outdir (str, optional): Defaults to &quot;.&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">taxid</span><span class="p">:</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="o">.</span><span class="n">taxid</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_loci" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_loci</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Generate a table of loci information.</p>


    <p><span class="doc-section-title">Yields:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(internal_locus_id, external_locus_id, [info])</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_loci</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a table of loci information.</span>

<span class="sd">    Yields:</span>
<span class="sd">        tuple: (internal_locus_id, external_locus_id, [info])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">_av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">locus</span> <span class="ow">in</span> <span class="n">_av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="nb">tuple</span><span class="p">(</span>
                <span class="p">[</span><span class="n">locus</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span>
                <span class="o">+</span> <span class="p">[</span><span class="n">locus</span><span class="o">.</span><span class="n">external_id</span><span class="p">]</span>
                <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_locus_to_protein" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_locus_to_protein</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function <code>table_locus_to_protein</code> generates a table of protein information from each locus, for import into Neo4j.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_locus_to_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function `table_locus_to_protein` generates a table of protein information from each locus, for import into Neo4j.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">ak</span><span class="p">,</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">loci</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">temp_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">loci</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
            <span class="c1"># sort features by id then start to maintain consistent output</span>
            <span class="n">temp_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;uid&quot;</span><span class="p">))</span>
            <span class="n">temp_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;start&quot;</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">temp_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">feature_is_protein</span><span class="p">():</span>
                    <span class="k">yield</span> <span class="p">(</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">locus_tag</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">description</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">partial_on_complete_genome</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">missing_start</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">missing_stop</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">internal_stop</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">partial_in_the_middle_of_a_contig</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">missing_N_terminus</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">missing_C_terminus</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">frameshifted</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">too_short_partial_abutting_assembly_gap</span><span class="p">,</span>
                        <span class="n">feature</span><span class="o">.</span><span class="n">incomplete</span><span class="p">,</span>
                    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_protein_ids" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_protein_ids</span><span class="p">(</span><span class="n">include_sequences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Protein hash id table for import into Neo4j</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_protein_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_sequences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Protein hash id table for import into Neo4j&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">include_sequences</span><span class="p">:</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="n">protein</span><span class="o">.</span><span class="n">crc64</span><span class="p">,</span> <span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="n">protein</span><span class="o">.</span><span class="n">crc64</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.table_protein_to_go" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">table_protein_to_go</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function <code>table_protein_to_go</code> iterates through the assemblies, loci, and features of a
given object, and yields tuples containing the protein hash and GO term for each feature that
has GO terms, for import into Neo4j.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">table_protein_to_go</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function `table_protein_to_go` iterates through the assemblies, loci, and features of a</span>
<span class="sd">    given object, and yields tuples containing the protein hash and GO term for each feature that</span>
<span class="sd">    has GO terms, for import into Neo4j.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">av</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">av</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="c1"># not all features will have goterms so check here</span>
                <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">goterms</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">goterm</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">goterms</span><span class="p">:</span>
                        <span class="k">yield</span> <span class="p">(</span>
                            <span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                            <span class="n">goterm</span><span class="o">.</span><span class="n">removeprefix</span><span class="p">(</span><span class="s2">&quot;GO:&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.write_fasta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">write_fasta</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="n">external_id</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>Write all proteins to a FASTA file</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outpath</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>path of file that FASTA entries will be appended to</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>external_id</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Write protein identifiers as the hash (True) or the original identifier (False). Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>**kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>see print(open_write.<strong>doc</strong>)</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_fasta</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">outpath</span><span class="p">,</span>
    <span class="n">external_id</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write all proteins to a FASTA file</span>

<span class="sd">    Args:</span>
<span class="sd">        outpath (str): path of file that FASTA entries will be appended to</span>
<span class="sd">        external_id (bool, optional): Write protein identifiers as the hash (True) or the original identifier (False). Defaults to False.</span>
<span class="sd">        **kwargs: see print(open_write.__doc__)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">filepath</span><span class="o">=</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">external_id</span><span class="p">:</span>
            <span class="n">fasta_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fasta_string_defline_external_id</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">fasta_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fasta_string_defline_uid</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fasta_gen</span><span class="p">:</span>
            <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="s2">&quot;compression&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;compression&quot;</span><span class="p">]:</span>
                <span class="n">handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">encode</span><span class="p">())</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">handle</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">counter</span><span class="p">)</span><span class="si">}</span><span class="s2"> proteins to </span><span class="si">{</span><span class="n">outpath</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.write_n_fasta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">write_n_fasta</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">n_splits</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function <code>write_n_fasta</code> exports protein sequences split into multiple fasta files.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outdir</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>outdir</code> parameter is a string that specifies the directory where the fasta files</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>will be saved.
  n_splits: The <code>n_splits</code> parameter in the <code>write_n_fasta</code> function determines the number of
fasta files the protein sequences will be split into. By default, it is set to 1, meaning all
the protein sequences will be written into a single fasta file. If you specify a value greater
than. Defaults to 1</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_n_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outdir</span><span class="p">,</span> <span class="n">n_splits</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function `write_n_fasta` exports protein sequences split into multiple fasta files.</span>

<span class="sd">    Args:</span>
<span class="sd">      outdir: The `outdir` parameter is a string that specifies the directory where the fasta files</span>
<span class="sd">    will be saved.</span>
<span class="sd">      n_splits: The `n_splits` parameter in the `write_n_fasta` function determines the number of</span>
<span class="sd">    fasta files the protein sequences will be split into. By default, it is set to 1, meaning all</span>
<span class="sd">    the protein sequences will be written into a single fasta file. If you specify a value greater</span>
<span class="sd">    than. Defaults to 1</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># this can be done with itertools.batched in python 3.12</span>
    <span class="k">def</span> <span class="nf">split</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
        <span class="c1"># https://stackoverflow.com/a/2135920</span>
        <span class="n">k</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">),</span> <span class="n">n</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">a</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">k</span> <span class="o">+</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">m</span><span class="p">)</span> <span class="p">:</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">k</span> <span class="o">+</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">m</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="n">protein_list</span> <span class="o">=</span> <span class="n">split</span><span class="p">(</span>
        <span class="p">[</span><span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span>
        <span class="n">n_splits</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;fasta_split_</span><span class="si">{</span><span class="n">counter</span><span class="si">}</span><span class="s2">.faa&quot;</span><span class="p">),</span> <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">protein_group</span><span class="p">:</span>
                <span class="n">handle</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h5 id="socialgene.base.socialgene.SocialGene.write_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">write_table</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">tablename</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">include_sequences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h5>


    <div class="doc doc-contents ">

        <p>The function writes a table to a specified output directory in TSV format, for import into Neo4j.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outdir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>outdir</code> parameter is a string that specifies the directory where the output</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>file will be saved.
  tablename (str): The <code>tablename</code> parameter is a string that specifies the name of the table to
be written.
  filename (str): The <code>filename</code> parameter is an optional argument that specifies the name of
the file to be written. If no <code>filename</code> is provided, the <code>tablename</code> will be used as the
filename.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/socialgene.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_table</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">outdir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">tablename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">include_sequences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function writes a table to a specified output directory in TSV format, for import into Neo4j.</span>

<span class="sd">    Args:</span>
<span class="sd">      outdir (str): The `outdir` parameter is a string that specifies the directory where the output</span>
<span class="sd">    file will be saved.</span>
<span class="sd">      tablename (str): The `tablename` parameter is a string that specifies the name of the table to</span>
<span class="sd">    be written.</span>
<span class="sd">      filename (str): The `filename` parameter is an optional argument that specifies the name of</span>
<span class="sd">    the file to be written. If no `filename` is provided, the `tablename` will be used as the</span>
<span class="sd">    filename.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">filename</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">tablename</span>
    <span class="n">outpath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">open_write</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
        <span class="n">tsv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span>
            <span class="n">handle</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_MINIMAL</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tablename</span><span class="p">)(</span><span class="n">include_sequences</span><span class="o">=</span><span class="n">include_sequences</span><span class="p">):</span>
            <span class="n">tsv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="socialgene.base.molbio"></a>
    <div class="doc doc-contents first">








  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Assembly" class="doc doc-heading">
            <code>Assembly</code>


</h3>


    <div class="doc doc-contents ">


        <p>Container class holding a dictionary of loci (ie genes/proteins)</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Assembly</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Container class holding a dictionary of loci (ie genes/proteins)&quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;parent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;loci&quot;</span><span class="p">,</span>
        <span class="s2">&quot;taxid&quot;</span><span class="p">,</span>
        <span class="s2">&quot;metadata&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uid&quot;</span><span class="p">,</span>
        <span class="s2">&quot;taxonomy&quot;</span><span class="p">,</span>
        <span class="s2">&quot;name&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">uid</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="n">parent</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="n">uid</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">loci</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxid</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxonomy</span> <span class="o">=</span> <span class="n">Taxonomy</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">LocusAssemblyMetadata</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">uid</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">feature_uid_set</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return all protein hashes within assembly</span>
<span class="sd">        Returns:</span>
<span class="sd">            set: protein hashes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">set</span><span class="p">()</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">feature_uid_set</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)}</span>

    <span class="k">def</span> <span class="nf">add_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">external_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a locus to an assembly object&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">external_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">external_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid4</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">external_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">external_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">Locus</span><span class="p">(</span><span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="n">external_id</span><span class="o">=</span><span class="n">external_id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">external_id</span><span class="si">}</span><span class="s2"> already present&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">fill_taxonomy_from_db</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
                <span class="n">res</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                        MATCH (a1:assembly {uid: $uid})-[:IS_TAXON]-&gt;(t2:taxid)-[:TAXON_PARENT*1..]-&gt;(t1:taxid)</span>
<span class="sd">                        WHERE t1.rank in [&quot;genus&quot;, &quot;family&quot;, &quot;order&quot;, &quot;class&quot;, &quot;phylum&quot;, &quot;clade&quot;, &quot;superkingdom&quot;]</span>
<span class="sd">                        WITH  apoc.map.fromLists([t2.rank],[t2.name]) as a, apoc.map.fromLists([t1.rank],[t1.name]) as b</span>
<span class="sd">                        return  apoc.map.mergeList(collect(a)+collect(b)) as tax_dict</span>
<span class="sd">                        &quot;&quot;&quot;</span><span class="p">,</span>
                    <span class="n">uid</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">value</span><span class="p">()</span>
                <span class="c1"># Dict comprehension appends &#39;_&#39; to the keys, to match the args in Taxonomy</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">taxonomy</span> <span class="o">=</span> <span class="n">Taxonomy</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">_&quot;</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()})</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error trying to retrieve taxonomy for </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">fill_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
                <span class="n">res</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                        MATCH (a1:assembly {uid: $uid})</span>
<span class="sd">                        return properties(a1)</span>
<span class="sd">                        &quot;&quot;&quot;</span><span class="p">,</span>
                    <span class="n">uid</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">value</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error trying to retrieve taxonomy for </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_locus_by_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">uid</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">uid</span> <span class="o">==</span> <span class="n">uid</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">v</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">gene_clusters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">locus</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">gene_cluster</span> <span class="ow">in</span> <span class="n">locus</span><span class="o">.</span><span class="n">gene_clusters</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">gene_cluster</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="n">v</span><span class="o">.</span><span class="n">fasta_string_defline_uid</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_external_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="n">v</span><span class="o">.</span><span class="n">fasta_string_defline_external_id</span>

    <span class="k">def</span> <span class="nf">_neo4j_node</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="n">n</span> <span class="o">=</span> <span class="n">ASSEMBLY_NODE</span><span class="p">()</span>
        <span class="n">dict_to_add</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;uid&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">}</span>
        <span class="n">dict_to_add</span> <span class="o">=</span> <span class="n">dict_to_add</span> <span class="o">|</span> <span class="p">{</span>
            <span class="n">k</span><span class="p">:</span> <span class="n">v</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ASSEMBLY_NODE</span><span class="o">.</span><span class="n">property_specification</span>
        <span class="p">}</span>
        <span class="n">n</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">(</span><span class="n">dict_to_add</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">n</span>

    <span class="k">def</span> <span class="nf">add_to_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">i</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">()</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxid</span><span class="p">:</span>
            <span class="n">taxnode</span> <span class="o">=</span> <span class="n">TAXID_NODE</span><span class="p">()</span>
            <span class="n">taxnode</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">({</span><span class="s2">&quot;uid&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxid</span><span class="p">})</span>
            <span class="n">IS_TAXON_REL</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">(),</span> <span class="n">end</span><span class="o">=</span><span class="n">taxnode</span><span class="p">)</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">(</span>
                <span class="n">create</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="socialgene.base.molbio.Assembly.feature_uid_set" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">feature_uid_set</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Return all protein hashes within assembly
Returns:
    set: protein hashes</p>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Assembly.add_locus" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">add_locus</span><span class="p">(</span><span class="n">external_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Add a locus to an assembly object</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">external_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a locus to an assembly object&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">external_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">external_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid4</span><span class="p">())</span>
    <span class="k">if</span> <span class="n">external_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">loci</span><span class="p">[</span><span class="n">external_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">Locus</span><span class="p">(</span><span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="n">external_id</span><span class="o">=</span><span class="n">external_id</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">external_id</span><span class="si">}</span><span class="s2"> already present&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Domain" class="doc doc-heading">
            <code>Domain</code>


</h3>


    <div class="doc doc-contents ">


        <p>Class for holding information about a domain/motif annotation</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Domain</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for holding information about a domain/motif annotation&quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;hmm_id&quot;</span><span class="p">,</span>
        <span class="s2">&quot;env_from&quot;</span><span class="p">,</span>  <span class="c1"># page 38 of the HMMER User Guide (http://eddylab.org/software/hmmer/Userguide.pdf) suggests using envelope (not hmm_from/to, ali_from/to)</span>
        <span class="s2">&quot;env_to&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq_pro_score&quot;</span><span class="p">,</span>
        <span class="s2">&quot;evalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;i_evalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;domain_bias&quot;</span><span class="p">,</span>
        <span class="s2">&quot;domain_score&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq_pro_bias&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hmm_from&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hmm_to&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ali_from&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ali_to&quot;</span><span class="p">,</span>
        <span class="s2">&quot;exponentialized&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">hmm_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">env_from</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">env_to</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">seq_pro_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">evalue</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">i_evalue</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">domain_bias</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">domain_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">seq_pro_bias</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">hmm_from</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">hmm_to</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ali_from</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ali_to</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">exponentialized</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>  <span class="c1"># this kwarg isn&#39;t accessed but is here so that calling Domain with dict unpacking with extra args doesn&#39;t fail</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Class for holding information about a domain/motif annotation</span>

<span class="sd">        Args:</span>
<span class="sd">          hmm_id (str): The `hmm_id` parameter is a string that represents the identifier of the hidden</span>
<span class="sd">        Markov model (HMM) associated with the domain.</span>
<span class="sd">          env_from (int): The `env_from` parameter represents the starting position of the domain in the</span>
<span class="sd">        target sequence.</span>
<span class="sd">          env_to (int): The `env_to` parameter represents the end position of the environment (sequence)</span>
<span class="sd">        in the domain. It is an integer value.</span>
<span class="sd">          seq_pro_score (float): The `seq_pro_score` parameter is a floating-point number that</span>
<span class="sd">        represents the sequence profile score of the domain.</span>
<span class="sd">          evalue (float): The `evalue` parameter is a floating-point number that represents the E-value</span>
<span class="sd">        of the domain. The E-value is a statistical measure that indicates the expected number of</span>
<span class="sd">        domains with a similar score or better that would occur by chance in a database of the same</span>
<span class="sd">        size.</span>
<span class="sd">          i_evalue (float): The `i_evalue` parameter is a floating-point number that represents the</span>
<span class="sd">        independent E-value of a domain. It is used to assess the statistical significance of the match</span>
<span class="sd">        between the domain and the sequence.</span>
<span class="sd">          domain_bias (float): The `domain_bias` parameter is a float that represents the bias score of</span>
<span class="sd">        a domain. It is used to measure the likelihood that a domain is present in a sequence due to</span>
<span class="sd">        chance rather than functional significance.</span>
<span class="sd">          domain_score (float): The `domain_score` parameter is a float that represents the score of the</span>
<span class="sd">        domain.</span>
<span class="sd">          seq_pro_bias (float): The `seq_pro_bias` parameter is a float value that represents the</span>
<span class="sd">        sequence profile bias of a domain. It is used to measure the bias in the sequence profile</span>
<span class="sd">        alignment.</span>
<span class="sd">          hmm_from (int): The `hmm_from` parameter represents the starting position of the domain in the</span>
<span class="sd">        HMM (Hidden Markov Model) profile. It is an integer value.</span>
<span class="sd">          hmm_to (int): The `hmm_to` parameter is an integer that represents the ending position of the</span>
<span class="sd">        hidden Markov model (HMM) alignment in the domain.</span>
<span class="sd">          ali_from (int): The `ali_from` parameter represents the starting position of the alignment in</span>
<span class="sd">        the sequence alignment. It is an integer value.</span>
<span class="sd">          ali_to (int): The `ali_to` parameter represents the ending position of the alignment in the</span>
<span class="sd">        sequence.</span>
<span class="sd">          exponentialized (bool): A boolean flag indicating whether the evalue and i_evalue should be</span>
<span class="sd">        exponentialized or not. If set to True, the evalue and i_evalue will be converted to exponential</span>
<span class="sd">        form. Defaults to True</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">Domain</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">exponentialized</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;exponentialized mus be bool, was </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">exponentialized</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exponentialized</span> <span class="o">=</span> <span class="n">exponentialized</span>
        <span class="k">if</span> <span class="n">exponentialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="n">find_exp</span><span class="p">(</span><span class="n">evalue</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">=</span> <span class="n">find_exp</span><span class="p">(</span><span class="n">i_evalue</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">evalue</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">i_evalue</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="c1"># some of are rounded because of differences between pyhmmer and hmmer results</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">hmm_id</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">env_from</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">env_from</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">env_to</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">env_to</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_score</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">seq_pro_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">domain_bias</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">domain_bias</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">domain_score</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">domain_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_bias</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">seq_pro_bias</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hmm_from</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hmm_from</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hmm_to</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hmm_to</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ali_from</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ali_from</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ali_to</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ali_to</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">ord_dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;hmm_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;env_from&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">env_from</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;env_to&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">env_to</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;seq_pro_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_score</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;evalue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">evalue</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;i_evalue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;domain_bias&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domain_bias</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;domain_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domain_score</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;seq_pro_bias&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_bias</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;hmm_from&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_from</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;hmm_to&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_to</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;ali_from&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ali_from</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;ali_to&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ali_to</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;exponentialized&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">bool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">exponentialized</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ord_dict</span>

    <span class="k">def</span> <span class="nf">tsv_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">ord_dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;hmm_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;env_from&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">env_from</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;env_to&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">env_to</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;seq_pro_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_score</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;evalue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">evalue</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;i_evalue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;domain_bias&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domain_bias</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;domain_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domain_score</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;seq_pro_bias&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_bias</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;hmm_from&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_from</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;hmm_to&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_to</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;ali_from&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ali_from</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;ali_to&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ali_to</span><span class="p">)</span>
        <span class="n">ord_dict</span><span class="p">[</span><span class="s2">&quot;exponentialized&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">bool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">exponentialized</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ord_dict</span>

    <span class="k">def</span> <span class="nf">get_hmm_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">domain_within_threshold</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if a protein&#39;s domain annotation is within the threshold for inclusion (currently i_evalue based)</span>
<span class="sd">        Returns:</span>
<span class="sd">            bool: is less than or equal to the set ievalue cutoff?</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">exponentialized</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">&lt;=</span> <span class="n">find_exp</span><span class="p">(</span><span class="n">env_vars</span><span class="p">[</span><span class="s2">&quot;HMMSEARCH_IEVALUE&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">&lt;=</span> <span class="n">env_vars</span><span class="p">[</span><span class="s2">&quot;HMMSEARCH_IEVALUE&quot;</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Used to prevent adding duplicate domains</span>
<span class="sd">        Returns:</span>
<span class="sd">            hash: hash for set()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">hash</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">env_from</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">env_to</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span><span class="p">))</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
            <span class="k">return</span> <span class="bp">NotImplemented</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">hmm_id</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">env_from</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">env_from</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">env_to</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">env_to</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">i_evalue</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="socialgene.base.molbio.Domain.domain_within_threshold" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">domain_within_threshold</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Check if a protein's domain annotation is within the threshold for inclusion (currently i_evalue based)
Returns:
    bool: is less than or equal to the set ievalue cutoff?</p>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Domain.__hash__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__hash__</span><span class="p">()</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Used to prevent adding duplicate domains
Returns:
    hash: hash for set()</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Used to prevent adding duplicate domains</span>
<span class="sd">    Returns:</span>
<span class="sd">        hash: hash for set()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">hash</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">env_from</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">env_to</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Domain.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">hmm_id</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">env_from</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">env_to</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seq_pro_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">i_evalue</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">domain_bias</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">domain_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seq_pro_bias</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">hmm_from</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">hmm_to</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ali_from</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ali_to</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">exponentialized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Class for holding information about a domain/motif annotation</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>hmm_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>hmm_id</code> parameter is a string that represents the identifier of the hidden</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Markov model (HMM) associated with the domain.
  env_from (int): The <code>env_from</code> parameter represents the starting position of the domain in the
target sequence.
  env_to (int): The <code>env_to</code> parameter represents the end position of the environment (sequence)
in the domain. It is an integer value.
  seq_pro_score (float): The <code>seq_pro_score</code> parameter is a floating-point number that
represents the sequence profile score of the domain.
  evalue (float): The <code>evalue</code> parameter is a floating-point number that represents the E-value
of the domain. The E-value is a statistical measure that indicates the expected number of
domains with a similar score or better that would occur by chance in a database of the same
size.
  i_evalue (float): The <code>i_evalue</code> parameter is a floating-point number that represents the
independent E-value of a domain. It is used to assess the statistical significance of the match
between the domain and the sequence.
  domain_bias (float): The <code>domain_bias</code> parameter is a float that represents the bias score of
a domain. It is used to measure the likelihood that a domain is present in a sequence due to
chance rather than functional significance.
  domain_score (float): The <code>domain_score</code> parameter is a float that represents the score of the
domain.
  seq_pro_bias (float): The <code>seq_pro_bias</code> parameter is a float value that represents the
sequence profile bias of a domain. It is used to measure the bias in the sequence profile
alignment.
  hmm_from (int): The <code>hmm_from</code> parameter represents the starting position of the domain in the
HMM (Hidden Markov Model) profile. It is an integer value.
  hmm_to (int): The <code>hmm_to</code> parameter is an integer that represents the ending position of the
hidden Markov model (HMM) alignment in the domain.
  ali_from (int): The <code>ali_from</code> parameter represents the starting position of the alignment in
the sequence alignment. It is an integer value.
  ali_to (int): The <code>ali_to</code> parameter represents the ending position of the alignment in the
sequence.
  exponentialized (bool): A boolean flag indicating whether the evalue and i_evalue should be
exponentialized or not. If set to True, the evalue and i_evalue will be converted to exponential
form. Defaults to True</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">hmm_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">env_from</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">env_to</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seq_pro_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">evalue</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">i_evalue</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">domain_bias</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">domain_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seq_pro_bias</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">hmm_from</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">hmm_to</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ali_from</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ali_to</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">exponentialized</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>  <span class="c1"># this kwarg isn&#39;t accessed but is here so that calling Domain with dict unpacking with extra args doesn&#39;t fail</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class for holding information about a domain/motif annotation</span>

<span class="sd">    Args:</span>
<span class="sd">      hmm_id (str): The `hmm_id` parameter is a string that represents the identifier of the hidden</span>
<span class="sd">    Markov model (HMM) associated with the domain.</span>
<span class="sd">      env_from (int): The `env_from` parameter represents the starting position of the domain in the</span>
<span class="sd">    target sequence.</span>
<span class="sd">      env_to (int): The `env_to` parameter represents the end position of the environment (sequence)</span>
<span class="sd">    in the domain. It is an integer value.</span>
<span class="sd">      seq_pro_score (float): The `seq_pro_score` parameter is a floating-point number that</span>
<span class="sd">    represents the sequence profile score of the domain.</span>
<span class="sd">      evalue (float): The `evalue` parameter is a floating-point number that represents the E-value</span>
<span class="sd">    of the domain. The E-value is a statistical measure that indicates the expected number of</span>
<span class="sd">    domains with a similar score or better that would occur by chance in a database of the same</span>
<span class="sd">    size.</span>
<span class="sd">      i_evalue (float): The `i_evalue` parameter is a floating-point number that represents the</span>
<span class="sd">    independent E-value of a domain. It is used to assess the statistical significance of the match</span>
<span class="sd">    between the domain and the sequence.</span>
<span class="sd">      domain_bias (float): The `domain_bias` parameter is a float that represents the bias score of</span>
<span class="sd">    a domain. It is used to measure the likelihood that a domain is present in a sequence due to</span>
<span class="sd">    chance rather than functional significance.</span>
<span class="sd">      domain_score (float): The `domain_score` parameter is a float that represents the score of the</span>
<span class="sd">    domain.</span>
<span class="sd">      seq_pro_bias (float): The `seq_pro_bias` parameter is a float value that represents the</span>
<span class="sd">    sequence profile bias of a domain. It is used to measure the bias in the sequence profile</span>
<span class="sd">    alignment.</span>
<span class="sd">      hmm_from (int): The `hmm_from` parameter represents the starting position of the domain in the</span>
<span class="sd">    HMM (Hidden Markov Model) profile. It is an integer value.</span>
<span class="sd">      hmm_to (int): The `hmm_to` parameter is an integer that represents the ending position of the</span>
<span class="sd">    hidden Markov model (HMM) alignment in the domain.</span>
<span class="sd">      ali_from (int): The `ali_from` parameter represents the starting position of the alignment in</span>
<span class="sd">    the sequence alignment. It is an integer value.</span>
<span class="sd">      ali_to (int): The `ali_to` parameter represents the ending position of the alignment in the</span>
<span class="sd">    sequence.</span>
<span class="sd">      exponentialized (bool): A boolean flag indicating whether the evalue and i_evalue should be</span>
<span class="sd">    exponentialized or not. If set to True, the evalue and i_evalue will be converted to exponential</span>
<span class="sd">    form. Defaults to True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">(</span><span class="n">Domain</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">exponentialized</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;exponentialized mus be bool, was </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">exponentialized</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">exponentialized</span> <span class="o">=</span> <span class="n">exponentialized</span>
    <span class="k">if</span> <span class="n">exponentialized</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="n">find_exp</span><span class="p">(</span><span class="n">evalue</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">=</span> <span class="n">find_exp</span><span class="p">(</span><span class="n">i_evalue</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">evalue</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">i_evalue</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">i_evalue</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
    <span class="c1"># some of are rounded because of differences between pyhmmer and hmmer results</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">hmm_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">hmm_id</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">env_from</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">env_from</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">env_to</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">env_to</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_score</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">seq_pro_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">domain_bias</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">domain_bias</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">domain_score</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">domain_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">seq_pro_bias</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">seq_pro_bias</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">hmm_from</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hmm_from</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">hmm_to</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hmm_to</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ali_from</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ali_from</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ali_to</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ali_to</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Feature" class="doc doc-heading">
            <code>Feature</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="socialgene.base.molbio.Location" href="#socialgene.base.molbio.Location">Location</a></code></p>


        <p>Container class for describing a feature on a locus</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Feature</span><span class="p">(</span><span class="n">Location</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Container class for describing a feature on a locus&quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;parent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uid&quot;</span><span class="p">,</span>
        <span class="s2">&quot;external_id&quot;</span><span class="p">,</span>
        <span class="s2">&quot;type&quot;</span><span class="p">,</span>
        <span class="s2">&quot;locus_tag&quot;</span><span class="p">,</span>
        <span class="s2">&quot;description&quot;</span><span class="p">,</span>
        <span class="s2">&quot;note&quot;</span><span class="p">,</span>
        <span class="s2">&quot;goterms&quot;</span><span class="p">,</span>
        <span class="s2">&quot;partial_on_complete_genome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;missing_start&quot;</span><span class="p">,</span>
        <span class="s2">&quot;missing_stop&quot;</span><span class="p">,</span>
        <span class="s2">&quot;internal_stop&quot;</span><span class="p">,</span>
        <span class="s2">&quot;partial_in_the_middle_of_a_contig&quot;</span><span class="p">,</span>
        <span class="s2">&quot;missing_N_terminus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;missing_C_terminus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;frameshifted&quot;</span><span class="p">,</span>
        <span class="s2">&quot;too_short_partial_abutting_assembly_gap&quot;</span><span class="p">,</span>
        <span class="s2">&quot;incomplete&quot;</span><span class="p">,</span>
        <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">uid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">external_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">locus_tag</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">note</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">goterms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">partial_on_complete_genome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">missing_start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">missing_stop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">internal_stop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">partial_in_the_middle_of_a_contig</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">missing_N_terminus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">missing_C_terminus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">frameshifted</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">too_short_partial_abutting_assembly_gap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">incomplete</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The above function is a constructor for a class that represents a feature on a locus, with</span>
<span class="sd">        various attributes and optional arguments.</span>

<span class="sd">        Args:</span>
<span class="sd">          uid (str): A string representing the hash value of the protein.</span>
<span class="sd">          external_id (str): The unique identifier for the protein associated with the feature.</span>
<span class="sd">          type (str): The &quot;type&quot; parameter is used to specify the type of feature. In this case, it is</span>
<span class="sd">        used to specify the type of the feature on a locus, such as &quot;protein&quot;.</span>
<span class="sd">          locus_tag (str): The locus_tag parameter is a string that represents the unique identifier for</span>
<span class="sd">        a feature on a locus. It is typically used in genomics to identify a specific gene or protein</span>
<span class="sd">        within a genome.</span>
<span class="sd">          description: A description of the feature on a locus.</span>
<span class="sd">          note: The `note` parameter is used to provide additional information or comments about the</span>
<span class="sd">        feature. It is an optional parameter and can be any string value.</span>
<span class="sd">          goterms: The `goterms` parameter is used to store the Gene Ontology terms associated with the</span>
<span class="sd">        feature. Gene Ontology (GO) terms are standardized terms used to describe the function,</span>
<span class="sd">        location, and involvement of genes and gene products in biological processes.</span>
<span class="sd">          partial_on_complete_genome: A boolean flag indicating whether the feature is partial on a</span>
<span class="sd">        complete genome.</span>
<span class="sd">          missing_start: A boolean indicating whether the start of the feature is missing.</span>
<span class="sd">          missing_stop: A boolean flag indicating whether the stop codon of the feature is missing.</span>
<span class="sd">          internal_stop: A boolean flag indicating whether the feature has an internal stop codon.</span>
<span class="sd">          partial_in_the_middle_of_a_contig: This parameter is used to indicate whether the feature is</span>
<span class="sd">        partial and located in the middle of a contig.</span>
<span class="sd">          missing_N_terminus: This parameter is used to indicate whether the N-terminus (the starting</span>
<span class="sd">        end) of the feature is missing. It is a boolean value, where True indicates that the N-terminus</span>
<span class="sd">        is missing and False indicates that it is not missing.</span>
<span class="sd">          missing_C_terminus: The parameter &quot;missing_C_terminus&quot; is used to indicate whether the feature</span>
<span class="sd">        is missing the C-terminus (the end) of the protein sequence. It is a boolean value that can be</span>
<span class="sd">        set to True or False.</span>
<span class="sd">          frameshifted: The &quot;frameshifted&quot; parameter is a boolean flag that indicates whether the</span>
<span class="sd">        feature has a frameshift mutation. A frameshift mutation occurs when the reading frame of a gene</span>
<span class="sd">        is disrupted by the insertion or deletion of nucleotides, causing a shift in the codon reading</span>
<span class="sd">        frame and potentially altering the amino</span>
<span class="sd">          too_short_partial_abutting_assembly_gap: This parameter is used to indicate whether the</span>
<span class="sd">        feature is too short and abuts an assembly gap.</span>
<span class="sd">          incomplete: A boolean flag indicating whether the feature is incomplete.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="n">parent</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="n">uid</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">external_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="nb">type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">locus_tag</span> <span class="o">=</span> <span class="n">locus_tag</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">description</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">note</span> <span class="o">=</span> <span class="n">note</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">goterms</span> <span class="o">=</span> <span class="n">goterms</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partial_on_complete_genome</span> <span class="o">=</span> <span class="n">partial_on_complete_genome</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missing_start</span> <span class="o">=</span> <span class="n">missing_start</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missing_stop</span> <span class="o">=</span> <span class="n">missing_stop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">internal_stop</span> <span class="o">=</span> <span class="n">internal_stop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partial_in_the_middle_of_a_contig</span> <span class="o">=</span> <span class="n">partial_in_the_middle_of_a_contig</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missing_N_terminus</span> <span class="o">=</span> <span class="n">missing_N_terminus</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missing_C_terminus</span> <span class="o">=</span> <span class="n">missing_C_terminus</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">frameshifted</span> <span class="o">=</span> <span class="n">frameshifted</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">too_short_partial_abutting_assembly_gap</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">too_short_partial_abutting_assembly_gap</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">incomplete</span> <span class="o">=</span> <span class="n">incomplete</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_is_protein</span><span class="p">():</span>
            <span class="c1"># Check if there is a SocialGene object in the parent chain</span>
            <span class="c1"># If there is, link the feature protein attribute to the SocialGene protein object</span>
            <span class="n">sg</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">current_object</span> <span class="o">=</span> <span class="bp">self</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">):</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">current_object</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__module__</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">current_object</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="o">==</span> <span class="s2">&quot;socialgene.base.socialgene.SocialGene&quot;</span>
                <span class="p">):</span>
                    <span class="n">sg</span> <span class="o">=</span> <span class="n">current_object</span>
                    <span class="k">break</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">current_object</span><span class="p">,</span> <span class="s2">&quot;parent&quot;</span><span class="p">):</span>
                        <span class="n">current_object</span> <span class="o">=</span> <span class="n">current_object</span><span class="o">.</span><span class="n">parent</span>
            <span class="k">if</span> <span class="n">sg</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="ow">in</span> <span class="n">sg</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)}</span>

    <span class="k">def</span> <span class="nf">feature_is_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if the feature &quot;is a protein&quot;</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: True if socialgene thinks it&#39;s a protein, False if not</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># types of genbank file features to classify as a &quot;protein&quot;</span>
        <span class="n">types_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="s2">&quot;CDS&quot;</span><span class="p">]</span>
        <span class="k">return</span> <span class="nb">any</span><span class="p">([</span><span class="kc">True</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">types_list</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">])</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="p">:</span>
                <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Erorr in fasta_string_defline_uid for </span><span class="si">{</span><span class="bp">self</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_external_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="p">:</span>
                <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Erorr in fasta_string_defline_external_id for </span><span class="si">{</span><span class="bp">self</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Used to prevent adding duplicate features to a locus (for hash in set() in Assembly.add_locus())</span>

<span class="sd">        Returns:</span>
<span class="sd">            hash: hash for set()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">hash</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Used for set() in Assembly.add_locus()&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
            <span class="k">return</span> <span class="bp">NotImplemented</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">parent</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">end</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">start</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">uid</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">strand</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">type</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="fm">__lt__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">&lt;</span> <span class="n">other</span><span class="o">.</span><span class="n">start</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">parent</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Feature.__eq__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__eq__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Used for set() in Assembly.add_locus()</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Used for set() in Assembly.add_locus()&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">parent</span>
        <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">end</span>
        <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">start</span>
        <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">uid</span>
        <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">strand</span>
        <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">type</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Feature.__hash__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__hash__</span><span class="p">()</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Used to prevent adding duplicate features to a locus (for hash in set() in Assembly.add_locus())</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>hash</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>hash for set()</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Used to prevent adding duplicate features to a locus (for hash in set() in Assembly.add_locus())</span>

<span class="sd">    Returns:</span>
<span class="sd">        hash: hash for set()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">hash</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Feature.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">uid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">external_id</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">locus_tag</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">note</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">goterms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">partial_on_complete_genome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">missing_start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">missing_stop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">internal_stop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">partial_in_the_middle_of_a_contig</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">missing_N_terminus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">missing_C_terminus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">frameshifted</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">too_short_partial_abutting_assembly_gap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">incomplete</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>The above function is a constructor for a class that represents a feature on a locus, with
various attributes and optional arguments.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>uid</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string representing the hash value of the protein.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>external_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The unique identifier for the protein associated with the feature.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>type</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The "type" parameter is used to specify the type of feature. In this case, it is</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>used to specify the type of the feature on a locus, such as "protein".
  locus_tag (str): The locus_tag parameter is a string that represents the unique identifier for
a feature on a locus. It is typically used in genomics to identify a specific gene or protein
within a genome.
  description: A description of the feature on a locus.
  note: The <code>note</code> parameter is used to provide additional information or comments about the
feature. It is an optional parameter and can be any string value.
  goterms: The <code>goterms</code> parameter is used to store the Gene Ontology terms associated with the
feature. Gene Ontology (GO) terms are standardized terms used to describe the function,
location, and involvement of genes and gene products in biological processes.
  partial_on_complete_genome: A boolean flag indicating whether the feature is partial on a
complete genome.
  missing_start: A boolean indicating whether the start of the feature is missing.
  missing_stop: A boolean flag indicating whether the stop codon of the feature is missing.
  internal_stop: A boolean flag indicating whether the feature has an internal stop codon.
  partial_in_the_middle_of_a_contig: This parameter is used to indicate whether the feature is
partial and located in the middle of a contig.
  missing_N_terminus: This parameter is used to indicate whether the N-terminus (the starting
end) of the feature is missing. It is a boolean value, where True indicates that the N-terminus
is missing and False indicates that it is not missing.
  missing_C_terminus: The parameter "missing_C_terminus" is used to indicate whether the feature
is missing the C-terminus (the end) of the protein sequence. It is a boolean value that can be
set to True or False.
  frameshifted: The "frameshifted" parameter is a boolean flag that indicates whether the
feature has a frameshift mutation. A frameshift mutation occurs when the reading frame of a gene
is disrupted by the insertion or deletion of nucleotides, causing a shift in the codon reading
frame and potentially altering the amino
  too_short_partial_abutting_assembly_gap: This parameter is used to indicate whether the
feature is too short and abuts an assembly gap.
  incomplete: A boolean flag indicating whether the feature is incomplete.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">uid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">external_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="nb">type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">locus_tag</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">note</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">goterms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">partial_on_complete_genome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">missing_start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">missing_stop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">internal_stop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">partial_in_the_middle_of_a_contig</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">missing_N_terminus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">missing_C_terminus</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">frameshifted</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">too_short_partial_abutting_assembly_gap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">incomplete</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The above function is a constructor for a class that represents a feature on a locus, with</span>
<span class="sd">    various attributes and optional arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">      uid (str): A string representing the hash value of the protein.</span>
<span class="sd">      external_id (str): The unique identifier for the protein associated with the feature.</span>
<span class="sd">      type (str): The &quot;type&quot; parameter is used to specify the type of feature. In this case, it is</span>
<span class="sd">    used to specify the type of the feature on a locus, such as &quot;protein&quot;.</span>
<span class="sd">      locus_tag (str): The locus_tag parameter is a string that represents the unique identifier for</span>
<span class="sd">    a feature on a locus. It is typically used in genomics to identify a specific gene or protein</span>
<span class="sd">    within a genome.</span>
<span class="sd">      description: A description of the feature on a locus.</span>
<span class="sd">      note: The `note` parameter is used to provide additional information or comments about the</span>
<span class="sd">    feature. It is an optional parameter and can be any string value.</span>
<span class="sd">      goterms: The `goterms` parameter is used to store the Gene Ontology terms associated with the</span>
<span class="sd">    feature. Gene Ontology (GO) terms are standardized terms used to describe the function,</span>
<span class="sd">    location, and involvement of genes and gene products in biological processes.</span>
<span class="sd">      partial_on_complete_genome: A boolean flag indicating whether the feature is partial on a</span>
<span class="sd">    complete genome.</span>
<span class="sd">      missing_start: A boolean indicating whether the start of the feature is missing.</span>
<span class="sd">      missing_stop: A boolean flag indicating whether the stop codon of the feature is missing.</span>
<span class="sd">      internal_stop: A boolean flag indicating whether the feature has an internal stop codon.</span>
<span class="sd">      partial_in_the_middle_of_a_contig: This parameter is used to indicate whether the feature is</span>
<span class="sd">    partial and located in the middle of a contig.</span>
<span class="sd">      missing_N_terminus: This parameter is used to indicate whether the N-terminus (the starting</span>
<span class="sd">    end) of the feature is missing. It is a boolean value, where True indicates that the N-terminus</span>
<span class="sd">    is missing and False indicates that it is not missing.</span>
<span class="sd">      missing_C_terminus: The parameter &quot;missing_C_terminus&quot; is used to indicate whether the feature</span>
<span class="sd">    is missing the C-terminus (the end) of the protein sequence. It is a boolean value that can be</span>
<span class="sd">    set to True or False.</span>
<span class="sd">      frameshifted: The &quot;frameshifted&quot; parameter is a boolean flag that indicates whether the</span>
<span class="sd">    feature has a frameshift mutation. A frameshift mutation occurs when the reading frame of a gene</span>
<span class="sd">    is disrupted by the insertion or deletion of nucleotides, causing a shift in the codon reading</span>
<span class="sd">    frame and potentially altering the amino</span>
<span class="sd">      too_short_partial_abutting_assembly_gap: This parameter is used to indicate whether the</span>
<span class="sd">    feature is too short and abuts an assembly gap.</span>
<span class="sd">      incomplete: A boolean flag indicating whether the feature is incomplete.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="n">parent</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="n">uid</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">external_id</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="nb">type</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">locus_tag</span> <span class="o">=</span> <span class="n">locus_tag</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">description</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">note</span> <span class="o">=</span> <span class="n">note</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">goterms</span> <span class="o">=</span> <span class="n">goterms</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">partial_on_complete_genome</span> <span class="o">=</span> <span class="n">partial_on_complete_genome</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">missing_start</span> <span class="o">=</span> <span class="n">missing_start</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">missing_stop</span> <span class="o">=</span> <span class="n">missing_stop</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">internal_stop</span> <span class="o">=</span> <span class="n">internal_stop</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">partial_in_the_middle_of_a_contig</span> <span class="o">=</span> <span class="n">partial_in_the_middle_of_a_contig</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">missing_N_terminus</span> <span class="o">=</span> <span class="n">missing_N_terminus</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">missing_C_terminus</span> <span class="o">=</span> <span class="n">missing_C_terminus</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">frameshifted</span> <span class="o">=</span> <span class="n">frameshifted</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">too_short_partial_abutting_assembly_gap</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">too_short_partial_abutting_assembly_gap</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">incomplete</span> <span class="o">=</span> <span class="n">incomplete</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_is_protein</span><span class="p">():</span>
        <span class="c1"># Check if there is a SocialGene object in the parent chain</span>
        <span class="c1"># If there is, link the feature protein attribute to the SocialGene protein object</span>
        <span class="n">sg</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">current_object</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">current_object</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__module__</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">current_object</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="o">==</span> <span class="s2">&quot;socialgene.base.socialgene.SocialGene&quot;</span>
            <span class="p">):</span>
                <span class="n">sg</span> <span class="o">=</span> <span class="n">current_object</span>
                <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">current_object</span><span class="p">,</span> <span class="s2">&quot;parent&quot;</span><span class="p">):</span>
                    <span class="n">current_object</span> <span class="o">=</span> <span class="n">current_object</span><span class="o">.</span><span class="n">parent</span>
        <span class="k">if</span> <span class="n">sg</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="ow">in</span> <span class="n">sg</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Feature.feature_is_protein" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">feature_is_protein</span><span class="p">()</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Check if the feature "is a protein"</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if socialgene thinks it's a protein, False if not</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">feature_is_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if the feature &quot;is a protein&quot;</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if socialgene thinks it&#39;s a protein, False if not</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># types of genbank file features to classify as a &quot;protein&quot;</span>
    <span class="n">types_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="s2">&quot;CDS&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">any</span><span class="p">([</span><span class="kc">True</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">types_list</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.FeatureCollection" class="doc doc-heading">
            <code>FeatureCollection</code>


</h3>


    <div class="doc doc-contents ">







              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">FeatureCollection</span><span class="p">:</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;parent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;features&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="nf">add_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a feature to a locus&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">Feature</span><span class="p">(</span><span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)}</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">feature_uid_set</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">}</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">features_sorted_by_midpoint</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sorts features by mid-coordinate of each feature&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="nb">int</span><span class="p">((</span><span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)):</span>
            <span class="k">yield</span> <span class="n">i</span>

    <span class="k">def</span> <span class="nf">drop_non_protein_features</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;protein&quot;</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">get_feature_by_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">uid</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">uid</span> <span class="o">==</span> <span class="n">uid</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">v</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">sg</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">current_object</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">current_object</span><span class="p">))</span>
                <span class="o">==</span> <span class="s2">&quot;&lt;class &#39;socialgene.base.socialgene.SocialGene&#39;&gt;&quot;</span>
            <span class="p">):</span>
                <span class="n">sg</span> <span class="o">=</span> <span class="n">current_object</span>
                <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">current_object</span> <span class="o">=</span> <span class="n">current_object</span><span class="o">.</span><span class="n">parent</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="n">i</span><span class="p">:</span> <span class="n">sg</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">}</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sg</span><span class="o">.</span><span class="n">proteins</span>
        <span class="p">}</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">protein_iter</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="n">i</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_external_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">yield</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="k">def</span> <span class="nf">_neo4j_node</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">n</span> <span class="o">=</span> <span class="n">NUCLEOTIDE_NODE</span><span class="p">()</span>
        <span class="n">dict_to_add</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;uid&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="s2">&quot;external_id&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="p">}</span>
        <span class="n">dict_to_add</span> <span class="o">=</span> <span class="n">dict_to_add</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">())</span>
        <span class="n">n</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">(</span><span class="n">dict_to_add</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">n</span>

    <span class="k">def</span> <span class="nf">add_to_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># create node</span>
        <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">()</span>
        <span class="n">n</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">(</span><span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">protein_nodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">encodes_rels</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="c1"># connect to proteins</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="n">protein_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">())</span>
            <span class="n">temp_rel</span> <span class="o">=</span> <span class="n">ENCODES_REL</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">())</span>
            <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span><span class="p">:</span> <span class="n">v</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ENCODES_REL</span><span class="o">.</span><span class="n">property_specification</span>
            <span class="p">}</span>
            <span class="n">properties</span> <span class="o">=</span> <span class="n">properties</span> <span class="o">|</span> <span class="p">{</span>
                <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
            <span class="p">}</span>
            <span class="n">temp_rel</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">(</span><span class="n">properties</span><span class="p">)</span>
            <span class="n">encodes_rels</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">temp_rel</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">protein_nodes</span><span class="p">:</span>
            <span class="n">protein_nodes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">protein_nodes</span><span class="p">)</span>
            <span class="n">protein_nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">add_multiple_to_neo4j</span><span class="p">(</span><span class="n">protein_nodes</span><span class="p">,</span> <span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">encodes_rels</span><span class="p">:</span>
            <span class="n">encodes_rels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">encodes_rels</span><span class="p">)</span>
            <span class="n">encodes_rels</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">add_multiple_to_neo4j</span><span class="p">(</span><span class="n">encodes_rels</span><span class="p">,</span> <span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="c1"># connect to assembly</span>
        <span class="n">assembly_node</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">()</span>
        <span class="n">assembly_node</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">(</span><span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ASSEMBLES_TO_REL</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">assembly_node</span><span class="p">)</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">(</span><span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">write_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="s2">&quot;-inf&quot;</span><span class="p">),</span> <span class="n">end</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">)):</span>
        <span class="n">record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
            <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">),</span>
            <span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="s2">&quot;A GenBank file generated by SocialGene.&quot;</span><span class="p">,</span>
            <span class="n">dbxrefs</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Assembly:</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="c1"># Add annotation</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features_sorted_by_midpoint</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">):</span>
                    <span class="k">continue</span>
            <span class="k">if</span> <span class="n">end</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">):</span>
                    <span class="k">continue</span>
            <span class="n">biofeat</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span>
                <span class="n">FeatureLocation</span><span class="p">(</span>
                    <span class="n">start</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span>
                    <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># biopython modifies the start position by 1</span>
                    <span class="n">end</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                    <span class="n">strand</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="nb">type</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
                <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span>
                    <span class="n">k</span><span class="p">:</span> <span class="n">v</span>
                    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">v</span> <span class="ow">and</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;parent&quot;</span>
                <span class="p">}</span>
                <span class="o">|</span> <span class="p">{</span><span class="s2">&quot;translation&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="p">},</span>
            <span class="p">)</span>
            <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">biofeat</span><span class="p">)</span>
        <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;DNA&quot;</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="n">record</span><span class="p">,</span>
            <span class="n">outpath</span><span class="p">,</span>
            <span class="s2">&quot;genbank&quot;</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="socialgene.base.molbio.FeatureCollection.features_sorted_by_midpoint" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">features_sorted_by_midpoint</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Sorts features by mid-coordinate of each feature</p>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.FeatureCollection.add_feature" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">add_feature</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Add a feature to a locus</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a feature to a locus&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">Feature</span><span class="p">(</span><span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Location" class="doc doc-heading">
            <code>Location</code>


</h3>


    <div class="doc doc-contents ">







              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Location</span><span class="p">:</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]</span>
    <span class="c1"># TODO: handle zero indexing</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">strand</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>  <span class="c1"># this kwarg isn&#39;t accessed but is here so that calling Location with dict unpacking with extra args doesn&#39;t fail</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Class describing genomic coordinates and strand direction.</span>

<span class="sd">        Args:</span>
<span class="sd">            start (int, optional): start coordinate</span>
<span class="sd">            end (int, optional): end coordinate</span>
<span class="sd">            strand (int, optional): DNA strand</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)}</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Location.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Class describing genomic coordinates and strand direction.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>start</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>start coordinate</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>end</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>end coordinate</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>strand</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DNA strand</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">strand</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>  <span class="c1"># this kwarg isn&#39;t accessed but is here so that calling Location with dict unpacking with extra args doesn&#39;t fail</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class describing genomic coordinates and strand direction.</span>

<span class="sd">    Args:</span>
<span class="sd">        start (int, optional): start coordinate</span>
<span class="sd">        end (int, optional): end coordinate</span>
<span class="sd">        strand (int, optional): DNA strand</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Locus" class="doc doc-heading">
            <code>Locus</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="socialgene.base.molbio.FeatureCollection" href="#socialgene.base.molbio.FeatureCollection">FeatureCollection</a></code></p>


        <p>Container holding a set() of genomic features</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Locus</span><span class="p">(</span><span class="n">FeatureCollection</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Container holding a set() of genomic features&quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;parent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;features&quot;</span><span class="p">,</span>
        <span class="s2">&quot;metadata&quot;</span><span class="p">,</span>
        <span class="s2">&quot;external_id&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uid&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gene_clusters&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">external_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="n">parent</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">LocusAssemblyMetadata</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">external_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calc_uid</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_clusters</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">calc_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">hasher</span><span class="o">.</span><span class="n">hasher</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">___</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_bgcs_by_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">features</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">Feature</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">features</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;All features must be of type Feature, not </span><span class="si">{</span><span class="p">[</span><span class="nb">type</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">i</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">Feature</span><span class="p">)]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">GeneCluster</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">add_bgcs_by_start_end</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">features</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="o">&lt;=</span> <span class="n">end</span><span class="p">}</span>
        <span class="k">if</span> <span class="n">features</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_bgcs_by_feature</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="n">features</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">write_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outpath</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
            <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">),</span>
            <span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="s2">&quot;A GenBank file generated by SocialGene.&quot;</span><span class="p">,</span>
            <span class="n">dbxrefs</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Assembly:</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="c1"># Add annotation</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features_sorted_by_midpoint</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">):</span>
                    <span class="k">continue</span>
            <span class="k">if</span> <span class="n">end</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">):</span>
                    <span class="k">continue</span>
            <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;protein&quot;</span><span class="p">:</span>
                <span class="n">feat_type</span> <span class="o">=</span> <span class="s2">&quot;CDS&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">feat_type</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span>
            <span class="n">biofeat</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span>
                <span class="n">FeatureLocation</span><span class="p">(</span>
                    <span class="n">start</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                    <span class="n">end</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                    <span class="n">strand</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="nb">type</span><span class="o">=</span><span class="n">feat_type</span><span class="p">,</span>
                <span class="n">qualifiers</span><span class="o">=</span><span class="p">{</span>
                    <span class="n">k</span><span class="p">:</span> <span class="n">v</span>
                    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">v</span> <span class="ow">and</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;parent&quot;</span>
                <span class="p">}</span>
                <span class="o">|</span> <span class="p">{</span><span class="s2">&quot;translation&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span><span class="p">},</span>
            <span class="p">)</span>
            <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">biofeat</span><span class="p">)</span>
        <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;DNA&quot;</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="n">record</span><span class="p">,</span>
            <span class="n">outpath</span><span class="p">,</span>
            <span class="s2">&quot;genbank&quot;</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Molbio" class="doc doc-heading">
            <code>Molbio</code>


</h3>


    <div class="doc doc-contents ">


        <p>Class for inheriting by SocialGene()</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Molbio</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for inheriting by SocialGene()&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>  <span class="c1"># TODO: ?</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># TODO: ?</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># TODO: ?</span>

    <span class="k">def</span> <span class="nf">get_all_feature_uids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a list of all proteins uids</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of all proteins uids</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>

    <span class="k">def</span> <span class="nf">add_protein</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">return_uid</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a protein to the protein dictionary</span>

<span class="sd">        Args:</span>
<span class="sd">            no_return (bool, optional): Whether the function should return the protein&#39;s uid. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Protein&#39;s hash</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">temp_protein</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="c1"># only add protein if it doesn&#39;t already exist</span>
        <span class="k">if</span> <span class="n">temp_protein</span><span class="o">.</span><span class="n">uid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
            <span class="c1"># deepcopy teo ensure instances aren&#39;t shared</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">temp_protein</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp_protein</span>
        <span class="k">if</span> <span class="n">return_uid</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">temp_protein</span><span class="o">.</span><span class="n">uid</span>

    <span class="k">def</span> <span class="nf">add_assembly</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">uid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add an assembly to a SocialGene object</span>

<span class="sd">        Args:</span>
<span class="sd">            uid (str, optional): Assembly identifier, should be unique across parsed input. Defaults to None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">uid</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">uid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid4</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">uid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">uid</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">parent</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uid</span><span class="si">}</span><span class="s2"> already present&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Molbio.add_assembly" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">add_assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Add an assembly to a SocialGene object</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>uid</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly identifier, should be unique across parsed input. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_assembly</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">uid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add an assembly to a SocialGene object</span>

<span class="sd">    Args:</span>
<span class="sd">        uid (str, optional): Assembly identifier, should be unique across parsed input. Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">uid</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">uid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid4</span><span class="p">())</span>
    <span class="k">if</span> <span class="n">uid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assemblies</span><span class="p">[</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">(</span><span class="n">uid</span><span class="o">=</span><span class="n">uid</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">parent</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uid</span><span class="si">}</span><span class="s2"> already present&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Molbio.add_protein" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">add_protein</span><span class="p">(</span><span class="n">return_uid</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Add a protein to the protein dictionary</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>no_return</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether the function should return the protein's uid. Defaults to False.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Protein's hash</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_protein</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">return_uid</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a protein to the protein dictionary</span>

<span class="sd">    Args:</span>
<span class="sd">        no_return (bool, optional): Whether the function should return the protein&#39;s uid. Defaults to False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Protein&#39;s hash</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">temp_protein</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="c1"># only add protein if it doesn&#39;t already exist</span>
    <span class="k">if</span> <span class="n">temp_protein</span><span class="o">.</span><span class="n">uid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
        <span class="c1"># deepcopy teo ensure instances aren&#39;t shared</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">temp_protein</span><span class="o">.</span><span class="n">uid</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp_protein</span>
    <span class="k">if</span> <span class="n">return_uid</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">temp_protein</span><span class="o">.</span><span class="n">uid</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Molbio.get_all_feature_uids" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_all_feature_uids</span><span class="p">()</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Return a list of all proteins uids</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of all proteins uids</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_all_feature_uids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a list of all proteins uids</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of all proteins uids</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">uid</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Protein" class="doc doc-heading">
            <code>Protein</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="socialgene.base.molbio.ProteinSequence" href="#socialgene.base.molbio.ProteinSequence">ProteinSequence</a></code></p>


        <p>Container class for describing a single protein</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Protein</span><span class="p">(</span>
    <span class="n">ProteinSequence</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Container class for describing a single protein&quot;&quot;&quot;</span>

    <span class="c1"># seqlen is included as a variable so it can be provided (e.g. a sequence isn&#39;t provided and proteins are created manually)</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">,</span> <span class="s2">&quot;external_id&quot;</span><span class="p">,</span> <span class="s2">&quot;domains&quot;</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">description</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">external_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">domains</span><span class="p">:</span> <span class="nb">set</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="o">*</span><span class="n">args</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;_summary_</span>

<span class="sd">        Args:</span>
<span class="sd">            description (str, optional): Protein description.</span>
<span class="sd">            external_id (str, optional): Non-hash-id descriptor (usually a database accession, e.g. NCBI&#39;s).</span>
<span class="sd">            seqlen (int, optional): Amino acid sequence length</span>
<span class="sd">            domains (Set, optional): Set of Domain() objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">description</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">external_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">domains</span> <span class="o">=</span> <span class="n">domains</span> <span class="k">if</span> <span class="n">domains</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">set</span><span class="p">()</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
            <span class="k">return</span> <span class="bp">NotImplemented</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">uid</span>

    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">hash</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
        <span class="p">}</span> <span class="o">|</span> <span class="p">{</span><span class="s2">&quot;seq_len&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_len</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">add_domain</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="n">args</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a domain to a protein if it meets threshold (ievalue)</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: used for counting # of domains added to protein</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">Domain</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">add_domains_from_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    MATCH (p1:protein)&lt;-[a1:ANNOTATES]-(h1:hmm)</span>
<span class="sd">                    where p1.uid in [&quot;_xvHFg1H1f43WxPcZT1P8Qbcx60chSxh&quot;]</span>
<span class="sd">                    RETURN apoc.map.merge({hmm_id:h1.uid}, properties(a1))</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
                    <span class="n">uid</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">),</span>
                <span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">add_domain</span><span class="p">(</span><span class="o">**</span><span class="n">i</span><span class="o">.</span><span class="n">value</span><span class="p">())</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error trying to retrieve domains for </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_sequences_from_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">GraphDriver</span><span class="p">()</span> <span class="k">as</span> <span class="n">db</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                    MATCH (p1:protein {uid: $uid})</span>
<span class="sd">                    RETURN p1.sequence as sequence</span>
<span class="sd">                    &quot;&quot;&quot;</span><span class="p">,</span>
                    <span class="n">uid</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">),</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="n">x</span> <span class="o">:=</span> <span class="n">i</span><span class="o">.</span><span class="n">value</span><span class="p">():</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">x</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error trying to retrieve sequences for </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">domain_list_sorted_by_mean_envelope_position</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># (ie can&#39;t sort a set())</span>
        <span class="c1"># &#39;if&#39; is to check whether domains is None</span>
        <span class="c1"># A tuple is passed so that the HMM annnotations are first sorted by the envelope midpoint</span>
        <span class="c1"># then ties are sorted alphabetically based on hmm_id for consistency</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">),</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">tx</span><span class="p">:</span> <span class="p">((</span><span class="n">tx</span><span class="o">.</span><span class="n">env_from</span> <span class="o">+</span> <span class="n">tx</span><span class="o">.</span><span class="n">env_to</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">tx</span><span class="o">.</span><span class="n">hmm_id</span><span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">domain_vector</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the domain uids for a protein as an ordered list</span>

<span class="sd">        list: list of domain uids</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tried to get domains from domain-less protein </span><span class="si">{</span><span class="bp">self</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">[]</span>
        <span class="k">return</span> <span class="p">[</span>
            <span class="n">i</span><span class="o">.</span><span class="n">get_hmm_id</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">domain_list_sorted_by_mean_envelope_position</span>
        <span class="p">]</span>

    <span class="k">def</span> <span class="nf">filter_domains</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prune all domains in all proteins that don&#39;t meet the inclusion threshold (currently HMMER&#39;s i_evalue)&quot;&quot;&quot;</span>
        <span class="n">_before_count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">)</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">domains</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">domains</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">domain_within_threshold</span><span class="p">}</span>

        <span class="k">del</span> <span class="n">temp</span>
        <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Removed </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">_before_count</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">))</span><span class="si">}</span><span class="s2"> domains from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_uid</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">fasta_string_defline_external_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="k">def</span> <span class="nf">_neo4j_node</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_sequences</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">PROTEIN_NODE</span><span class="p">:</span>
        <span class="n">n</span> <span class="o">=</span> <span class="n">PROTEIN_NODE</span><span class="p">()</span>
        <span class="n">dict_to_add</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;uid&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">uid</span><span class="p">,</span> <span class="s2">&quot;crc64&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">crc64</span><span class="p">}</span>
        <span class="k">if</span> <span class="n">include_sequences</span><span class="p">:</span>
            <span class="n">dict_to_add</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span>
        <span class="n">n</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">(</span><span class="n">dict_to_add</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">n</span>

    <span class="k">def</span> <span class="nf">add_to_neo4j</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_sequences</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">(</span><span class="n">include_sequences</span><span class="o">=</span><span class="n">include_sequences</span><span class="p">)</span>
        <span class="n">n</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">(</span><span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">:</span>
            <span class="n">hmmnode</span> <span class="o">=</span> <span class="n">HMM_NODE</span><span class="p">()</span>
            <span class="n">hmmnode</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">({</span><span class="s2">&quot;uid&quot;</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">hmm_id</span><span class="p">})</span>
            <span class="n">rel</span> <span class="o">=</span> <span class="n">ANNOTATES_REL</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">hmmnode</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_neo4j_node</span><span class="p">())</span>
            <span class="n">rel</span><span class="o">.</span><span class="n">fill_from_dict</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="n">k</span><span class="p">:</span> <span class="n">v</span>
                    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">all_attributes</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ANNOTATES_REL</span><span class="o">.</span><span class="n">property_specification</span>
                <span class="p">}</span>
            <span class="p">)</span>
            <span class="n">rel</span><span class="o">.</span><span class="n">add_to_neo4j</span><span class="p">(</span><span class="n">create</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="socialgene.base.molbio.Protein.domain_vector" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">domain_vector</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Get the domain uids for a protein as an ordered list</p>
<p>list: list of domain uids</p>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Protein.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">external_id</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">domains</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p><em>summary</em></p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>description</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Protein description.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>external_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Non-hash-id descriptor (usually a database accession, e.g. NCBI's).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seqlen</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Amino acid sequence length</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>domains</code>
            </td>
            <td>
                  <code>Set</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Set of Domain() objects.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">description</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">external_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">domains</span><span class="p">:</span> <span class="nb">set</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="o">*</span><span class="n">args</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;_summary_</span>

<span class="sd">    Args:</span>
<span class="sd">        description (str, optional): Protein description.</span>
<span class="sd">        external_id (str, optional): Non-hash-id descriptor (usually a database accession, e.g. NCBI&#39;s).</span>
<span class="sd">        seqlen (int, optional): Amino acid sequence length</span>
<span class="sd">        domains (Set, optional): Set of Domain() objects.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">description</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">external_id</span> <span class="o">=</span> <span class="n">external_id</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">domains</span> <span class="o">=</span> <span class="n">domains</span> <span class="k">if</span> <span class="n">domains</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">set</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Protein.add_domain" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">add_domain</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Add a domain to a protein if it meets threshold (ievalue)</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>used for counting # of domains added to protein</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_domain</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="o">*</span><span class="n">args</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a domain to a protein if it meets threshold (ievalue)</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: used for counting # of domains added to protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">Domain</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.Protein.filter_domains" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_domains</span><span class="p">()</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Prune all domains in all proteins that don't meet the inclusion threshold (currently HMMER's i_evalue)</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_domains</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prune all domains in all proteins that don&#39;t meet the inclusion threshold (currently HMMER&#39;s i_evalue)&quot;&quot;&quot;</span>
    <span class="n">_before_count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">)</span>
    <span class="n">temp</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">domains</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">domains</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">domain_within_threshold</span><span class="p">}</span>

    <span class="k">del</span> <span class="n">temp</span>
    <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Removed </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">_before_count</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">domains</span><span class="p">))</span><span class="si">}</span><span class="s2"> domains from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">external_id</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.ProteinSequence" class="doc doc-heading">
            <code>ProteinSequence</code>


</h3>


    <div class="doc doc-contents ">


        <p>Class used for working with protein sequences and can be initialized with either a sequence or a uid.</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">ProteinSequence</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class used for working with protein sequences and can be initialized with either a sequence or a uid.&quot;&quot;&quot;</span>

    <span class="n">_amino_acids</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;N&quot;</span><span class="p">,</span>
        <span class="s2">&quot;D&quot;</span><span class="p">,</span>
        <span class="s2">&quot;C&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Q&quot;</span><span class="p">,</span>
        <span class="s2">&quot;E&quot;</span><span class="p">,</span>
        <span class="s2">&quot;G&quot;</span><span class="p">,</span>
        <span class="s2">&quot;H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;I&quot;</span><span class="p">,</span>
        <span class="s2">&quot;L&quot;</span><span class="p">,</span>
        <span class="s2">&quot;K&quot;</span><span class="p">,</span>
        <span class="s2">&quot;M&quot;</span><span class="p">,</span>
        <span class="s2">&quot;F&quot;</span><span class="p">,</span>
        <span class="s2">&quot;P&quot;</span><span class="p">,</span>
        <span class="s2">&quot;S&quot;</span><span class="p">,</span>
        <span class="s2">&quot;T&quot;</span><span class="p">,</span>
        <span class="s2">&quot;W&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Y&quot;</span><span class="p">,</span>
        <span class="s2">&quot;V&quot;</span><span class="p">,</span>
        <span class="s2">&quot;X&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Z&quot;</span><span class="p">,</span>
        <span class="s2">&quot;J&quot;</span><span class="p">,</span>
        <span class="s2">&quot;U&quot;</span><span class="p">,</span>
        <span class="s2">&quot;B&quot;</span><span class="p">,</span>
        <span class="s2">&quot;O&quot;</span><span class="p">,</span>
        <span class="s2">&quot;*&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;uid&quot;</span><span class="p">,</span> <span class="s2">&quot;__crc64&quot;</span><span class="p">,</span> <span class="s2">&quot;__md5&quot;</span><span class="p">,</span> <span class="s2">&quot;_seq_len&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence&quot;</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">uid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">seq_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Class for holding an amino acid sequence (protein)</span>

<span class="sd">        Args:</span>
<span class="sd">            sequence (str, optional): amino acid sequence</span>
<span class="sd">            uid (str, optional): a uid can be provided if a sequence isn&#39;t</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: Must provide either a sequence or a uid</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># the sequence input variable has a default so that domtblout can create a socialgene object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">)):</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">uid</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Must provide either a sequence or a uid&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="n">uid</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_assign_hash</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">seq_len</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_seq_len</span> <span class="o">=</span> <span class="n">seq_len</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">seq_len</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the length of the protein</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: number of amino acids</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="nd">@seq_len</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">seq_len</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_seq_len</span> <span class="o">=</span> <span class="n">value</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_seq_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function returns a dictionary containing the attributes of an object.</span>

<span class="sd">        Returns:</span>
<span class="sd">          The `__dict__` method is returning a dictionary that contains the names and values of all the</span>
<span class="sd">        attributes of the object. The attributes are obtained using the `getattr` function and are filtered</span>
<span class="sd">        to only include attributes that are defined in the `__slots__` list and that exist in the object.</span>
<span class="sd">        The dictionary is sorted based on the names of the attributes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)}</span>

    <span class="k">def</span> <span class="nf">_one_letter_amino_acids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create an ordered dictionary of amino acids. Used to count AAs in a protein sequence.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: amino acid count</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">OrderedDict</span><span class="p">({</span><span class="n">i</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_amino_acids</span><span class="p">})</span>

    <span class="k">def</span> <span class="nf">_amino_acid_count</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a &#39;-&#39; separated string of an amino acid count</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: sequence of amino acids</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;0&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_one_letter_amino_acids</span><span class="p">()])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_one_letter_amino_acids</span><span class="p">()]</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_assign_hash</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function assigns a hash value to a sequence of amino acids.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_standardize_sequence</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="n">hasher</span><span class="o">.</span><span class="n">hash_aminos</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">crc64</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">hasher</span><span class="o">.</span><span class="n">hash_aminos</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">algo</span><span class="o">=</span><span class="s2">&quot;crc64&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">md5</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">hasher</span><span class="o">.</span><span class="n">hash_aminos</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">algo</span><span class="o">=</span><span class="s2">&quot;md5&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_standardize_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The function converts the protein sequence to uppercase and checks if all characters are valid</span>
<span class="sd">        amino acids, raising an error if an unknown character is found.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_amino_acids</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)]):</span>
            <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown character/letter in protein sequence&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="socialgene.base.molbio.ProteinSequence.seq_len" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">seq_len</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
      <small class="doc doc-label doc-label-writable"><code>writable</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Return the length of the protein</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of amino acids</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.ProteinSequence.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">uid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seq_len</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>Class for holding an amino acid sequence (protein)</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sequence</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>amino acid sequence</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>uid</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>a uid can be provided if a sequence isn't</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Must provide either a sequence or a uid</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">uid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">seq_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for holding an amino acid sequence (protein)</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence (str, optional): amino acid sequence</span>
<span class="sd">        uid (str, optional): a uid can be provided if a sequence isn&#39;t</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: Must provide either a sequence or a uid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># the sequence input variable has a default so that domtblout can create a socialgene object</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">)):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">uid</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Must provide either a sequence or a uid&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">uid</span> <span class="o">=</span> <span class="n">uid</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_assign_hash</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">seq_len</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_seq_len</span> <span class="o">=</span> <span class="n">seq_len</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="socialgene.base.molbio.ProteinSequence.all_attributes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">all_attributes</span><span class="p">()</span></code>

</h4>


    <div class="doc doc-contents ">

        <p>The function returns a dictionary containing the attributes of an object.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>__dict__</code> method is returning a dictionary that contains the names and values of all the</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
        <p>attributes of the object. The attributes are obtained using the <code>getattr</code> function and are filtered
to only include attributes that are defined in the <code>__slots__</code> list and that exist in the object.
The dictionary is sorted based on the names of the attributes.</p>

            <details class="quote">
              <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">all_attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The function returns a dictionary containing the attributes of an object.</span>

<span class="sd">    Returns:</span>
<span class="sd">      The `__dict__` method is returning a dictionary that contains the names and values of all the</span>
<span class="sd">    attributes of the object. The attributes are obtained using the `getattr` function and are filtered</span>
<span class="sd">    to only include attributes that are defined in the `__slots__` list and that exist in the object.</span>
<span class="sd">    The dictionary is sorted based on the names of the attributes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__slots__</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">)}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="socialgene.base.molbio.Taxonomy" class="doc doc-heading">
            <code>Taxonomy</code>


</h3>


    <div class="doc doc-contents ">


        <p>Class is a reserved word so just underscore all ranks to be consistent</p>






              <details class="quote">
                <summary>Source code in <code>socialgene/base/molbio.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Taxonomy</span><span class="p">:</span>
    <span class="s2">&quot;Class is a reserved word so just underscore all ranks to be consistent&quot;</span>
    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;species_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genus_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;family_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;order_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;class_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;phylum_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;clade_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;superkingdom_&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">species_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">genus_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">family_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">order_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">class_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">phylum_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">clade_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">superkingdom_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_</span> <span class="o">=</span> <span class="n">species_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genus_</span> <span class="o">=</span> <span class="n">genus_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">family_</span> <span class="o">=</span> <span class="n">family_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">order_</span> <span class="o">=</span> <span class="n">order_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">class_</span> <span class="o">=</span> <span class="n">class_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phylum_</span> <span class="o">=</span> <span class="n">phylum_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clade_</span> <span class="o">=</span> <span class="n">clade_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">superkingdom_</span> <span class="o">=</span> <span class="n">superkingdom_</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div>




  </div>

    </div>

</div>









  



  <form class="md-feedback" name="feedback" hidden>
    <fieldset>
      <legend class="md-feedback__title">
        Was this page helpful?
      </legend>
      <div class="md-feedback__inner">
        <div class="md-feedback__list">
          
            <button class="md-feedback__icon md-icon" type="submit" title="This page was helpful" data-md-value="1">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg>
            </button>
          
            <button class="md-feedback__icon md-icon" type="submit" title="This page could be improved" data-md-value="0">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg>
            </button>
          
        </div>
        <div class="md-feedback__note">
          
            <div data-md-value="1" hidden>
              
              
                
              
              Thanks for your feedback!
            </div>
          
            <div data-md-value="0" hidden>
              
              
                
              
              Thanks for your feedback! Help us improve this page by using our <a href="..." target="_blank" rel="noopener">feedback form</a>.
            </div>
          
        </div>
      </div>
    </fieldset>
  </form>


                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["content.action.edit", "content.code.copy"], "search": "../../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.83f73b43.min.js"></script>
      
    
  </body>
</html>