<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.533">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>schema</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="schema_files/libs/clipboard/clipboard.min.js"></script>
<script src="schema_files/libs/quarto-html/quarto.js"></script>
<script src="schema_files/libs/quarto-html/popper.min.js"></script>
<script src="schema_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="schema_files/libs/quarto-html/anchor.min.js"></script>
<link href="schema_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="schema_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="schema_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="schema_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="schema_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<p>The schema below is auto-generatated via the <a href="https://github.com/socialgene/sgpy">SocialGene Python Package</a>.</p>
<pre class="shell"><code># pip install socialgene
sg_schema --nodes  --markdown
sg_schema --rels  --markdown</code></pre>
<section id="nodes" class="level2">
<h2 class="anchored" data-anchor-id="nodes">Nodes</h2>
<table class="table">
<colgroup>
<col style="width: 2%">
<col style="width: 16%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 73%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Label</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">NF results subdirectory</th>
<th style="text-align: left;">Neo4j header file</th>
<th style="text-align: left;">Unique on</th>
<th style="text-align: left;">properties</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">assembly</td>
<td style="text-align: left;">Represents a single genome/assembly/BGC. If the input was a FASTA file or if assembly wasn’t in the genbank metadata then this will represent the file the data came from.</td>
<td style="text-align: left;">genomic_info</td>
<td style="text-align: left;">assembly.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘altitude’, ‘bio_material’, ‘bioproject’, ‘biosample’, ‘cell_line’, ‘cell_type’, ‘chromosome’, ‘clone’, ‘clone_lib’, ‘collected_by’, ‘collection_date’, ‘country’, ‘cultivar’, ‘culture_collection’, ‘db_xref’, ‘dev_stage’, ‘ecotype’, ‘environmental_sample’, ‘focus’, ‘germline’, ‘haplogroup’, ‘haplotype’, ‘host’, ‘identified_by’, ‘isolate’, ‘isolation_source’, ‘lab_host’, ‘lat_lon’, ‘macronuclear’, ‘map’, ‘mating_type’, ‘metagenome_source’, ‘mol_type’, ‘note’, ‘organelle’, ‘organism’, ‘pcr_primers’, ‘plasmid’, ‘pop_variant’, ‘proviral’, ‘rearranged’, ‘segment’, ‘serotype’, ‘serovar’, ‘sex’, ‘specimen_voucher’, ‘strain’, ‘sub_clone’, ‘submitter_seqid’, ‘sub_species’, ‘sub_strain’, ‘tissue_lib’, ‘tissue_type’, ‘transgenic’, ‘type_material’, ‘variety’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">assembly</td>
<td style="text-align: left;">Represents a single Mibig entry</td>
<td style="text-align: left;">genomic_info</td>
<td style="text-align: left;">assembly.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chebi</td>
<td style="text-align: left;">Represents a ChEBI term</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">chembl</td>
<td style="text-align: left;">Represents a CHEMBL term</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chemical_compound:substructure</td>
<td style="text-align: left;">Represents a chemical substructure</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">inchi, CanonicalSmiles</td>
<td style="text-align: left;">[‘uid’, ‘MolWt’, ‘HeavyAtomMolWt’, ‘ExactMolWt’, ‘NumValenceElectrons’, ‘NumRadicalElectrons’, ‘HeavyAtomCount’, ‘NumAliphaticCarbocycles’, ‘NumAliphaticHeterocycles’, ‘NumAliphaticRings’, ‘NumAromaticCarbocycles’, ‘NumAromaticHeterocycles’, ‘NumAromaticRings’, ‘NumHAcceptors’, ‘NumHDonors’, ‘NumHeteroatoms’, ‘NumRotatableBonds’, ‘NumSaturatedCarbocycles’, ‘NumSaturatedHeterocycles’, ‘NumSaturatedRings’, ‘RingCount’, ‘MolLogP’, ‘MolMR’, ‘AnonymousGraph’, ‘ElementGraph’, ‘CanonicalSmiles’, ‘MurckoScaffold’, ‘ExtendedMurcko’, ‘MolFormula’, ‘AtomBondCounts’, ‘DegreeVector’, ‘Mesomer’, ‘HetAtomTautomer’, ‘HetAtomProtomer’, ‘RedoxPair’, ‘Regioisomer’, ‘NetCharge’, ‘SmallWorldIndexBR’, ‘SmallWorldIndexBRL’, ‘ArthorSubstructureOrder’, ‘HetAtomTautomerv2’, ‘inchi’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">chemical_compound</td>
<td style="text-align: left;">Represents a chemical compound</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">inchi, CanonicalSmiles</td>
<td style="text-align: left;">[‘uid’, ‘MolWt’, ‘HeavyAtomMolWt’, ‘ExactMolWt’, ‘NumValenceElectrons’, ‘NumRadicalElectrons’, ‘HeavyAtomCount’, ‘NumAliphaticCarbocycles’, ‘NumAliphaticHeterocycles’, ‘NumAliphaticRings’, ‘NumAromaticCarbocycles’, ‘NumAromaticHeterocycles’, ‘NumAromaticRings’, ‘NumHAcceptors’, ‘NumHDonors’, ‘NumHeteroatoms’, ‘NumRotatableBonds’, ‘NumSaturatedCarbocycles’, ‘NumSaturatedHeterocycles’, ‘NumSaturatedRings’, ‘RingCount’, ‘MolLogP’, ‘MolMR’, ‘AnonymousGraph’, ‘ElementGraph’, ‘CanonicalSmiles’, ‘MurckoScaffold’, ‘ExtendedMurcko’, ‘MolFormula’, ‘AtomBondCounts’, ‘DegreeVector’, ‘Mesomer’, ‘HetAtomTautomer’, ‘HetAtomProtomer’, ‘RedoxPair’, ‘Regioisomer’, ‘NetCharge’, ‘SmallWorldIndexBR’, ‘SmallWorldIndexBRL’, ‘ArthorSubstructureOrder’, ‘HetAtomTautomerv2’, ‘inchi’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chemical_compound:substrate</td>
<td style="text-align: left;">Mibig substrate (e.g.&nbsp;NRPS monomer)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">inchi, CanonicalSmiles</td>
<td style="text-align: left;">[‘uid’, ‘MolWt’, ‘HeavyAtomMolWt’, ‘ExactMolWt’, ‘NumValenceElectrons’, ‘NumRadicalElectrons’, ‘HeavyAtomCount’, ‘NumAliphaticCarbocycles’, ‘NumAliphaticHeterocycles’, ‘NumAliphaticRings’, ‘NumAromaticCarbocycles’, ‘NumAromaticHeterocycles’, ‘NumAromaticRings’, ‘NumHAcceptors’, ‘NumHDonors’, ‘NumHeteroatoms’, ‘NumRotatableBonds’, ‘NumSaturatedCarbocycles’, ‘NumSaturatedHeterocycles’, ‘NumSaturatedRings’, ‘RingCount’, ‘MolLogP’, ‘MolMR’, ‘AnonymousGraph’, ‘ElementGraph’, ‘CanonicalSmiles’, ‘MurckoScaffold’, ‘ExtendedMurcko’, ‘MolFormula’, ‘AtomBondCounts’, ‘DegreeVector’, ‘Mesomer’, ‘HetAtomTautomer’, ‘HetAtomProtomer’, ‘RedoxPair’, ‘Regioisomer’, ‘NetCharge’, ‘SmallWorldIndexBR’, ‘SmallWorldIndexBRL’, ‘ArthorSubstructureOrder’, ‘HetAtomTautomerv2’, ‘inchi’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">chemical_fragment</td>
<td style="text-align: left;">Represents a chemical fragment as defined by rdkit.Chem.Descriptors</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">classyfire</td>
<td style="text-align: left;">Represents a classyfire chemical ontology term</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘name’, ‘definition’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">gnps_cluster</td>
<td style="text-align: left;">Represents a GNPS molecular networking cluster</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">cluster_index, workflow_uuid</td>
<td style="text-align: left;">[‘workflow_uuid’, ‘defaultgroups’, ‘g1’, ‘g2’, ‘g3’, ‘g4’, ‘g5’, ‘g6’, ‘gnpslinkout_cluster’, ‘gnpslinkout_network’, ‘mqscore’, ‘mzerrorppm’, ‘massdiff’, ‘rtmean’, ‘rtmean_min’, ‘rtstderr’, ‘uniquefilesources’, ‘uniquefilesourcescount’, ‘cluster_index’, ‘componentindex’, ‘number_of_spectra’, ‘parent_mass’, ‘precursor_charge’, ‘precursor_mass’, ‘sumprecursor_intensity’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gnps_library_spectrum</td>
<td style="text-align: left;">Represents a GNPS library spectrum</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘compound_name’, ‘compound_source’, ‘pi’, ‘data_collector’, ‘adduct’, ‘precursor_mz’, ‘exactmass’, ‘charge’, ‘cas_number’, ‘pubmed_id’, ‘smiles’, ‘inchi’, ‘inchi_aux’, ‘library_class’, ‘ionmode’, ‘libraryqualitystring’, ‘mqscore’, ‘tic_query’, ‘rt_query’, ‘mzerrorppm’, ‘sharedpeaks’, ‘massdiff’, ‘libmz’, ‘specmz’, ‘speccharge’, ‘moleculeexplorerdatasets’, ‘moleculeexplorerfiles’, ‘molecular_formula’, ‘inchikey’, ‘inchikey_planar’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">gnps_organism</td>
<td style="text-align: left;">Represents an organism (as defined by GNPS)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">goterm</td>
<td style="text-align: left;">Represent a GO term</td>
<td style="text-align: left;">goterms</td>
<td style="text-align: left;">goterms.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘name’, ‘namespace’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">hmm</td>
<td style="text-align: left;">Represents a single non-redundant HMM model</td>
<td style="text-align: left;">hmm_info</td>
<td style="text-align: left;">sg_hmm_nodes.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">hmm_source</td>
<td style="text-align: left;">Represents the source of an HMM model (e.g.&nbsp;PFAM)</td>
<td style="text-align: left;">hmm_info</td>
<td style="text-align: left;">hmm_source.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘:LABEL’, ‘rel_path’, ‘name’, ‘acc’, ‘notes’, ‘description’, ‘date’, ‘hash’, ‘hash_used’, ‘model_length’, ‘super_category’, ‘category’, ‘subcategory’, ‘ga’, ‘tc’, ‘nc’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">instrument</td>
<td style="text-align: left;">Represents an instrument</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ion_source</td>
<td style="text-align: left;">Represents an ion source</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">npatlas</td>
<td style="text-align: left;">Represents a single NPAtlas entry</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘original_name’, ‘mol_formula’, ‘mol_weight’, ‘exact_mass’, ‘inchikey’, ‘smiles’, ‘cluster_id’, ‘node_id’, ‘synonyms’, ‘inchi’, ‘m_plus_h’, ‘m_plus_na’, ‘genus’, ‘species’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">npclassifier_class</td>
<td style="text-align: left;">Represents a NPClassifier class</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">npclassifier_pathway</td>
<td style="text-align: left;">Represents a NPClassifier pathway</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">npclassifier_superclass</td>
<td style="text-align: left;">Represents a NPClassifier superclass</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">npmrd</td>
<td style="text-align: left;">Represents a single NP-MRD entry</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">nucleotide</td>
<td style="text-align: left;">Represents a single nucleotide sequence (e.g.&nbsp;a contig/scaffold/chromosome)</td>
<td style="text-align: left;">genomic_info</td>
<td style="text-align: left;">locus.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘external_id’, ‘altitude’, ‘bio_material’, ‘bioproject’, ‘biosample’, ‘cell_line’, ‘cell_type’, ‘chromosome’, ‘clone’, ‘clone_lib’, ‘collected_by’, ‘collection_date’, ‘country’, ‘cultivar’, ‘culture_collection’, ‘db_xref’, ‘dev_stage’, ‘ecotype’, ‘environmental_sample’, ‘focus’, ‘germline’, ‘haplogroup’, ‘haplotype’, ‘host’, ‘identified_by’, ‘isolate’, ‘isolation_source’, ‘lab_host’, ‘lat_lon’, ‘macronuclear’, ‘map’, ‘mating_type’, ‘metagenome_source’, ‘mol_type’, ‘note’, ‘organelle’, ‘organism’, ‘pcr_primers’, ‘plasmid’, ‘pop_variant’, ‘proviral’, ‘rearranged’, ‘segment’, ‘serotype’, ‘serovar’, ‘sex’, ‘specimen_voucher’, ‘strain’, ‘sub_clone’, ‘submitter_seqid’, ‘sub_species’, ‘sub_strain’, ‘tissue_lib’, ‘tissue_type’, ‘transgenic’, ‘type_material’, ‘variety’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">parameters</td>
<td style="text-align: left;">Parameters and environmental variables used during database creation</td>
<td style="text-align: left;">parameters</td>
<td style="text-align: left;">parameters.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘SG_LOC_NEO4J’, ‘SG_LOC_HMMS’, ‘NEO4J_dbms_memory_pagecache_size’, ’NEO4J_dbms_memory_heap_initial__size’, ’NEO4J_dbms_memory_heap_max__size’, ‘HMMSEARCH_IEVALUE’, ‘HMMSEARCH_BACKGROUND’, ‘HMMSEARCH_BIASFILTER’, ‘HMMSEARCH_NULL2’, ‘HMMSEARCH_SEED’, ‘HMMSEARCH_Z’, ‘HMMSEARCH_DOMZ’, ‘HMMSEARCH_F1’, ‘HMMSEARCH_F2’, ‘HMMSEARCH_F3’, ‘HMMSEARCH_E’, ‘HMMSEARCH_DOME’, ‘HMMSEARCH_INCE’, ‘HMMSEARCH_INCDOME’, ‘HMMSEARCH_BITCUTOFFS’, ‘platform’, ‘architecture’, ‘py_executable’, ‘py_version’, ‘genome_download_command’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">protein</td>
<td style="text-align: left;">Represents a non-redundant protein</td>
<td style="text-align: left;">protein_info</td>
<td style="text-align: left;">protein_ids.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘crc64’, ‘sequence’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">publication</td>
<td style="text-align: left;">Represents a publication</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">doi</td>
<td style="text-align: left;">[‘doi’, ‘pmid’, ‘authors’, ‘title’, ‘journal’, ‘year’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">publication</td>
<td style="text-align: left;">Represents a publication</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">doi</td>
<td style="text-align: left;">[‘doi’, ‘pmid’, ‘authors’, ‘title’, ‘journal’, ‘year’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">spectrum</td>
<td style="text-align: left;">Represents a GNPS molecular networking spectrum</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘original_filename’, ‘parentmass’, ‘charge’, ‘rettime’, ‘assembly’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">taxid</td>
<td style="text-align: left;">Represents a single taxon within NCBI taxonomy</td>
<td style="text-align: left;">taxdump_process</td>
<td style="text-align: left;">taxid.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’, ‘name’, ‘rank’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">tigrfam_mainrole</td>
<td style="text-align: left;">Represents a TIGRFAM main role</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfam_mainrole.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tigrfam_role</td>
<td style="text-align: left;">Represents a TIGRFAM role</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfam_role.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
<tr class="even">
<td style="text-align: left;">tigrfam_subrole</td>
<td style="text-align: left;">Represents a TIGRFAM sub role</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfam_subrole.header</td>
<td style="text-align: left;">uid</td>
<td style="text-align: left;">[‘uid’]</td>
</tr>
</tbody>
</table>
</section>
<section id="relationships" class="level2">
<h2 class="anchored" data-anchor-id="relationships">Relationships</h2>
<table class="table">
<colgroup>
<col style="width: 14%">
<col style="width: 43%">
<col style="width: 17%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Label</th>
<th style="text-align: left;">Relationship</th>
<th style="text-align: left;">NF results subdirectory</th>
<th style="text-align: left;">Neo4j header file</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ALTERNATIVE_PARENTS</td>
<td style="text-align: left;">(npatlas)-[:ALTERNATIVE_PARENTS]-&gt;(classyfire)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">ASSEMBLES_TO</td>
<td style="text-align: left;">(nucleotide)-[:ASSEMBLES_TO]-&gt;(assembly)</td>
<td style="text-align: left;">genomic_info</td>
<td style="text-align: left;">assembly_to_locus.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ANNOTATES</td>
<td style="text-align: left;">(hmm)-[:ANNOTATES]-&gt;(protein)</td>
<td style="text-align: left;">parsed_domtblout</td>
<td style="text-align: left;">protein_to_hmm_header.header</td>
</tr>
<tr class="even">
<td style="text-align: left;">BLASTP</td>
<td style="text-align: left;">(protein)-[:BLASTP]-&gt;(protein)</td>
<td style="text-align: left;">diamond_blastp</td>
<td style="text-align: left;">blastp.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CONTAINS</td>
<td style="text-align: left;">(chemical_compound)-[:CONTAINS]-&gt;(chemical_fragment)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">DIRECT_PARENT</td>
<td style="text-align: left;">(npatlas)-[:DIRECT_PARENT]-&gt;(classyfire)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENCODES</td>
<td style="text-align: left;">(nucleotide)-[:ENCODES]-&gt;(protein)</td>
<td style="text-align: left;">genomic_info</td>
<td style="text-align: left;">locus_to_protein.header</td>
</tr>
<tr class="even">
<td style="text-align: left;">FROM</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:FROM]-&gt;(instrument)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FROM</td>
<td style="text-align: left;">(gnps_cluster)-[:FROM]-&gt;(assembly)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">FROM</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:FROM]-&gt;(ion_source)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FROM</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:FROM]-&gt;(gnps_organism)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO_ANN</td>
<td style="text-align: left;">(hmm_source)-[:GO_ANN]-&gt;(goterm)</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfam_to_go.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GOTERM_RELS</td>
<td style="text-align: left;">(goterm)-[:GOTERM_RELS]-&gt;(goterm)</td>
<td style="text-align: left;">goterms</td>
<td style="text-align: left;">go_to_go.header</td>
</tr>
<tr class="even">
<td style="text-align: left;">HAS</td>
<td style="text-align: left;">(npatlas)-[:HAS]-&gt;(publication)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HAS</td>
<td style="text-align: left;">(npatlas)-[:HAS]-&gt;(gnps_library_spectrum)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">HAS</td>
<td style="text-align: left;">(npatlas)-[:HAS]-&gt;(npmrd)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(npatlas)-[:IS_A]-&gt;(npclassifier_pathway)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(classyfire)-[:IS_A]-&gt;(classyfire)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(npatlas)-[:IS_A]-&gt;(npclassifier_class)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:IS_A]-&gt;(npclassifier_pathway)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">IS_TAXON</td>
<td style="text-align: left;">(assembly)-[:IS_TAXON]-&gt;(taxid)</td>
<td style="text-align: left;">genomic_info</td>
<td style="text-align: left;">assembly_to_taxid.header</td>
</tr>
<tr class="even">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:IS_A]-&gt;(npclassifier_superclass)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:IS_A]-&gt;(npclassifier_class)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(gnps_library_spectrum)-[:IS_A]-&gt;(chemical_compound)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(npatlas)-[:IS_A]-&gt;(chemical_compound)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">INTERMEDIATE_NODES</td>
<td style="text-align: left;">(npatlas)-[:INTERMEDIATE_NODES]-&gt;(classyfire)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">IS_A</td>
<td style="text-align: left;">(npatlas)-[:IS_A]-&gt;(npclassifier_superclass)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">LIBRARY_HIT</td>
<td style="text-align: left;">(gnps_cluster)-[:LIBRARY_HIT]-&gt;(gnps_library_spectrum)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LOWEST_CLASS</td>
<td style="text-align: left;">(npatlas)-[:LOWEST_CLASS]-&gt;(classyfire)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">MMSEQS2</td>
<td style="text-align: left;">(protein)-[:MMSEQS2]-&gt;(protein)</td>
<td style="text-align: left;">mmseqs2_cluster</td>
<td style="text-align: left;">mmseqs2.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">MOLECULAR_NETWORK</td>
<td style="text-align: left;">(gnps_cluster)-[:MOLECULAR_NETWORK]-&gt;(gnps_cluster)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">MAINROLE_ANN</td>
<td style="text-align: left;">(tigrfam_role)-[:MAINROLE_ANN]-&gt;(tigrfam_mainrole)</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfamrole_to_mainrole.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PRODUCES</td>
<td style="text-align: left;">(taxid)-[:PRODUCES]-&gt;(npatlas)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">PRODUCES</td>
<td style="text-align: left;">(assembly)-[:PRODUCES]-&gt;(npatlas)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PROTEIN_TO_GO</td>
<td style="text-align: left;">(protein)-[:PROTEIN_TO_GO]-&gt;(goterm)</td>
<td style="text-align: left;">protein_info</td>
<td style="text-align: left;">protein_to_go.header</td>
</tr>
<tr class="even">
<td style="text-align: left;">ROLE_ANN</td>
<td style="text-align: left;">(hmm_source)-[:ROLE_ANN]-&gt;(tigrfam_role)</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfam_to_role.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SYNONYM</td>
<td style="text-align: left;">(classyfire)-[:SYNONYM]-&gt;(chebi)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">SOURCE_DB</td>
<td style="text-align: left;">(hmm)-[:SOURCE_DB]-&gt;(hmm_source)</td>
<td style="text-align: left;">hmm_info</td>
<td style="text-align: left;">hmm_source_relationships.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SIMILAR</td>
<td style="text-align: left;">(chemical_compound)-[:SIMILAR]-&gt;(chemical_compound)</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">None</td>
</tr>
<tr class="even">
<td style="text-align: left;">SUBROLE_ANN</td>
<td style="text-align: left;">(tigrfam_role)-[:SUBROLE_ANN]-&gt;(tigrfam_subrole)</td>
<td style="text-align: left;">tigrfam_info</td>
<td style="text-align: left;">tigrfamrole_to_subrole.header</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TAXON_PARENT</td>
<td style="text-align: left;">(taxid)-[:TAXON_PARENT]-&gt;(taxid)</td>
<td style="text-align: left;">taxdump_process</td>
<td style="text-align: left;">taxid_to_taxid.header</td>
</tr>
</tbody>
</table>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>